<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>GEO DataSets: Experiment Selection and Initial Processing</title>

  <meta property="description" itemprop="description" content="A walkthrough of processing bulk RNA-seq data from the GEO database purely in R."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-07-05"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-07-05"/>
  <meta name="article:author" content="Feifei Li"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="GEO DataSets: Experiment Selection and Initial Processing"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="A walkthrough of processing bulk RNA-seq data from the GEO database purely in R."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="GEO DataSets: Experiment Selection and Initial Processing"/>
  <meta property="twitter:description" content="A walkthrough of processing bulk RNA-seq data from the GEO database purely in R."/>

  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=GEOmetadb: powerful alternative search engine for the Gene Expression Omnibus;citation_publication_date=2008;citation_volume=24;citation_doi=10.1093/bioinformatics/btn520;citation_issn=1367-4803;citation_author=Yuelin Zhu;citation_author=Sean Davis;citation_author=Robert Stephens;citation_author=Paul S. Meltzer;citation_author=Yidong Chen"/>
  <meta name="citation_reference" content="citation_title=GEOquery: a bridge between the Gene Expression Omnibus (GEO) and BioConductor;citation_publication_date=2007;citation_volume=23;citation_doi=10.1093/bioinformatics/btm254;citation_issn=1367-4803;citation_author=Sean Davis;citation_author=Paul S. Meltzer"/>
  <meta name="citation_reference" content="citation_title=edgeR: a Bioconductor package for differential expression analysis of digital gene expression data;citation_publication_date=2009;citation_volume=26;citation_doi=10.1093/bioinformatics/btp616;citation_issn=1367-4803;citation_author=Mark D. Robinson;citation_author=Davis J. McCarthy;citation_author=Gordon K. Smyth"/>
  <meta name="citation_reference" content="citation_title=Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation;citation_publication_date=2012;citation_volume=40;citation_doi=10.1093/nar/gks042;citation_issn=0305-1048;citation_author=Davis J. McCarthy;citation_author=Yunshun Chen;citation_author=Gordon K. Smyth"/>
  <meta name="citation_reference" content="citation_title=Selecting between-sample RNA-seq normalization methods from the perspective of their assumptions;citation_publication_date=2016;citation_volume=19;citation_doi=10.1093/bib/bbx008;citation_author=Ciaran Evans;citation_author=Johanna Hardin;citation_author=Daniel Stoebel"/>
  <meta name="citation_reference" content="citation_title=A scaling normalization method for differential expression analysis of RNA-seq data;citation_publication_date=2010;citation_publisher=BioMed Central;citation_volume=11;citation_author=Mark D Robinson;citation_author=Alicia Oshlack"/>
  <meta name="citation_reference" content="citation_title=Transcriptomic similarities and differences in host response between SARS-CoV-2 and other viral infections;citation_publication_date=2021;citation_volume=24;citation_doi=https://doi.org/10.1016/j.isci.2020.101947;citation_issn=2589-0042;citation_author=Simone A. Thair;citation_author=Yudong D. He;citation_author=Yehudit Hasin-Brumshtein;citation_author=Suraj Sakaram;citation_author=Rushika Pandya;citation_author=Jiaying Toh;citation_author=David Rawling;citation_author=Melissa Remmel;citation_author=Sabrina Coyle;citation_author=George N. Dalekos;citation_author=Ioannis Koutsodimitropoulos;citation_author=Glykeria Vlachogianni;citation_author=Eleni Gkeka;citation_author=Eleni Karakike;citation_author=Georgia Damoraki;citation_author=Nikolaos Antonakos;citation_author=Purvesh Khatri;citation_author=Evangelos J. Giamarellos-Bourboulis;citation_author=Timothy E. Sweeney"/>
  <meta name="citation_reference" content="citation_title=Count-based differential expression analysis of RNA sequencing data using r and bioconductor;citation_publication_date=2013;citation_volume=8;citation_doi=10.1038/nprot.2013.099;citation_author=Simon Anders;citation_author=Davis Mccarthy;citation_author=Yunshun Chen;citation_author=Michal Okoniewski;citation_author=Gordon Smyth;citation_author=Wolfgang Huber;citation_author=Mark Robinson"/>
  <meta name="citation_reference" content="citation_title=Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation;citation_publication_date=2012;citation_volume=40;citation_doi=10.1093/nar/gks042;citation_issn=0305-1048;citation_author=Davis J. McCarthy;citation_author=Yunshun Chen;citation_author=Gordon K. Smyth"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","bibliography","output","creative_commons","categories","preview"]}},"value":[{"type":"character","attributes":{},"value":["GEO DataSets: Experiment Selection and Initial Processing"]},{"type":"character","attributes":{},"value":["A walkthrough of processing bulk RNA-seq data from the GEO database purely in R.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","url"]}},"value":[{"type":"character","attributes":{},"value":["Feifei Li"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":[]}},"value":[]}]}]},{"type":"character","attributes":{},"value":["2021-07-05"]},{"type":"character","attributes":{},"value":["biblio.bib"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["R","bioconductor","bioinformatics"]},{"type":"character","attributes":{},"value":["geo.gif"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["biblio.bib","data/DGEList.rds","data/GEOmetadb.sqlite","data/GSE152641_Inflammatix_COVID19_counts_entrez.csv.gz","data/normalised_counts.rds","data/samples.rds","geo-datasets-experiment-selection-and-initial-processing_files/anchor-4.2.2/anchor.min.js","geo-datasets-experiment-selection-and-initial-processing_files/bowser-1.9.3/bowser.min.js","geo-datasets-experiment-selection-and-initial-processing_files/distill-2.2.21/template.v2.js","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/boxplot-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/density-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-32-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-33-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-34-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-35-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-36-1.png","geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-37-1.png","geo-datasets-experiment-selection-and-initial-processing_files/header-attrs-2.8.6/header-attrs.js","geo-datasets-experiment-selection-and-initial-processing_files/header-attrs-2.9/header-attrs.js","geo-datasets-experiment-selection-and-initial-processing_files/jquery-1.11.3/jquery.min.js","geo-datasets-experiment-selection-and-initial-processing_files/kePrint-0.0.1/kePrint.js","geo-datasets-experiment-selection-and-initial-processing_files/lightable-0.0.1/lightable.css","geo-datasets-experiment-selection-and-initial-processing_files/popper-2.6.0/popper.min.js","geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy-bundle.umd.min.js","geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy-light-border.css","geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy.css","geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy.umd.min.js","geo-datasets-experiment-selection-and-initial-processing_files/webcomponents-2.0.0/webcomponents.js","geo.gif","geometadb_er.png"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          return "<p>" + $('#ref-' + ref).html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/header-attrs-2.9/header-attrs.js"></script>
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/kePrint-0.0.1/kePrint.js"></script>
  <link href="geo-datasets-experiment-selection-and-initial-processing_files/lightable-0.0.1/lightable.css" rel="stylesheet" />
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/jquery-1.11.3/jquery.min.js"></script>
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/popper-2.6.0/popper.min.js"></script>
  <link href="geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="geo-datasets-experiment-selection-and-initial-processing_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"GEO DataSets: Experiment Selection and Initial Processing","description":"A walkthrough of processing bulk RNA-seq data from the GEO database purely in R.","authors":[{"author":"Feifei Li","authorURL":{},"affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2021-07-05T00:00:00.000-04:00","citationText":"Li, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>GEO DataSets: Experiment Selection and Initial Processing</h1>
<!--radix_placeholder_categories-->
<div class="dt-tags">
<div class="dt=tag">R</div>
<div class="dt=tag">bioconductor</div>
<div class="dt=tag">bioinformatics</div>
</div>
<!--/radix_placeholder_categories-->
<p><p>A walkthrough of processing bulk RNA-seq data from the GEO database purely in R.</p></p>
</div>

<div class="d-byline">
  Feifei Li true 
  
<br/>2021-07-05
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#finding-experiments-of-interest">Finding Experiments of Interest</a>
<ul>
<li><a href="#geo-metadata-repository-geometadb">GEO metadata repository: GEOmetadb</a></li>
<li><a href="#querying-geo-database">Querying GEO database</a></li>
<li><a href="#filtering-query-result">Filtering Query Result</a></li>
<li><a href="#overview-of-the-chosen-experiment">Overview of the chosen experiment</a></li>
<li><a href="#downloading-dataset">Downloading Dataset</a></li>
</ul></li>
<li><a href="#rna-seq-counts-normalisation">RNA-seq Counts Normalisation</a>
<ul>
<li><a href="#tmm-normalisation">TMM Normalisation</a></li>
<li><a href="#distribution-for-read-counts---before-vs.-after-normalisation">Distribution for read counts - Before vs. after normalisation</a></li>
</ul></li>
<li><a href="#post-normalisation-analysis">Post-normalisation Analysis</a>
<ul>
<li><a href="#biological-coefficient-of-variation-bcv">Biological coefficient of variation (BCV)</a></li>
<li><a href="#mean-variance-relationship">Mean-variance relationship</a></li>
<li><a href="#sample-relations">Sample relations</a></li>
</ul></li>
<li><a href="#to-do">To Do</a></li>
</ul>
</nav>
</div>
<p>This article demonstrates a pure R workflow of finding experiments of interest, downloading datasets of experiments of interest from the Gene Expression Omnibus (GEO) database, and processing the bulk RNA-seq counts data.</p>
<p>The readers are assumed to have certain degree of understanding of what the GEO database is.</p>
<h2 id="finding-experiments-of-interest">Finding Experiments of Interest</h2>
<h3 id="geo-metadata-repository-geometadb">GEO metadata repository: GEOmetadb</h3>
<p><code>GEOmetadb</code> is a thin wrapper around a SQLite database into which all the most up-to-date NCBI GEO metadata that are associated with GEO samples (GSM), GEO platforms (GPL), GEO data series (GSE), and curated GEO datasets (GDS) are parsed.<span class="citation" data-cites="GEOmetadb">(<a href="#ref-GEOmetadb" role="doc-biblioref">Zhu et al. 2008</a>)</span> It enables querying the GEO database using R scripts without accessing the webpage. Although NCBI’s public API to the NCBI Entrez system <a href="https://www.ncbi.nlm.nih.gov/books/NBK25501/"><strong>The E-utilities</strong></a> also also offers an option for scripting queries for finding datasets of interest in the GEO database, the E-utilities are difficult to use (e.g. might need HTTP POST calls for long queries) and require extra effort to learn.</p>
<p>To install <code>GEOmetadb</code> via Bioconductor:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"GEOmetadb"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>
    <span class='fu'>BiocManager</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocManager/man/install.html'>install</a></span><span class='op'>(</span><span class='st'>"GEOmetadb"</span>, ask <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>We will also create a directory <code>./data</code> dedicated to storing data files under the working directory:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span><span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.exists</a></span><span class='op'>(</span><span class='st'>"./data"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='https://rdrr.io/r/base/files2.html'>dir.create</a></span><span class='op'>(</span><span class='st'>"./data"</span>, showWarnings <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</div>
<p>Then we download the SQLite database that has the GEO metadata into <code>./data</code> via <code>GEOmetadb</code>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Download GEOmeta database if not found</span>
<span class='va'>metadb</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"GEOmetadb.sqlite"</span><span class='op'>)</span>
<span class='kw'>if</span><span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/files.html'>file.exists</a></span><span class='op'>(</span><span class='va'>metadb</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'>## avoid re-downloading the 11G database everytime running the script</span>
    <span class='va'>metadb</span> <span class='op'>&lt;-</span> <span class='fu'>GEOmetadb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOmetadb/man/getSQLiteFile.html'>getSQLiteFile</a></span><span class='op'>(</span>destdir <span class='op'>=</span> <span class='st'>"./data"</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</div>
<h3 id="querying-geo-database">Querying GEO database</h3>
<p>For this demonstration, we are interested in experiments that:</p>
<ul>
<li>were done in the recent 6 years</li>
<li>are studies on COVID-19 (SARS-CoV-2)</li>
<li>used native human cells or tissues</li>
<li>have RNA-seq data</li>
</ul>
<p>To build an query to find experiments that meet the above conditions:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>query</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span>
    <span class='st'>"SELECT DISTINCT"</span>,
    <span class='st'>"gse.title,"</span>,
    <span class='st'>"gse.gse,"</span>,
    <span class='st'>"gpl.title AS platform,"</span>,
    <span class='st'>"gse.submission_date,"</span>,
    <span class='st'>"gse.supplementary_file"</span>,
    <span class='st'>"FROM"</span>,
    <span class='st'>"gse"</span>,
    <span class='st'>"JOIN gse_gpl ON gse_gpl.gse = gse.gse"</span>,
    <span class='st'>"JOIN gpl ON gse_gpl.gpl = gpl.gpl"</span>,
    <span class='st'>"WHERE"</span>,
    <span class='st'>"gse.submission_date &gt; '2015-01-01' AND"</span>,           <span class='co'># not older than 6 years</span>
    <span class='st'>"gse.title LIKE '%SARS-CoV-2%' AND "</span>,                  <span class='co'># experiment on COVID</span>
    <span class='st'>"gpl.organism LIKE '%Homo sapiens%' AND"</span>, <span class='co'># dataset of human cells or tissue</span>
    <span class='st'>"gpl.technology LIKE '%high-throughput seq%'"</span><span class='op'>)</span>                <span class='co'># RNA-seq data</span>
</code></pre>
</div>
</div>
<p>The ER diagram for the GEOmetadb SQLite database is represented as the following. This might come in useful when constructing an SQL statement querying the database.</p>
<figure>
<img src="geometadb_er.png" alt="The entity-relationship diagram for the GEOmetadb database (Zhu et al. 2008)" /><figcaption aria-hidden="true">The entity-relationship diagram for the GEOmetadb database <span class="citation" data-cites="GEOmetadb">(<a href="#ref-GEOmetadb" role="doc-biblioref">Zhu et al. 2008</a>)</span></figcaption>
</figure>
<p>For the GEO accession prefixes present in the figure:</p>
<ul>
<li><code>GSE</code>: Series
<ul>
<li>An original submitter-supplied record that summarizes an experiment</li>
<li>links together a group of related Samples and provides a focal point and description of the whole study</li>
<li>may also contain tables describing extracted data, summary conclusions, or analyses</li>
</ul></li>
<li><code>GSM</code>: Sample
<ul>
<li>describes the conditions under which an individual Sample was handled, the manipulations it underwent, and the abundance measurement of each element derived from it</li>
</ul></li>
<li><code>GPL</code>: Platform
<ul>
<li>composed of a summary description of the array or sequencer and, for array-based Platforms, a data table defining the array template</li>
<li>may reference many Samples that have been submitted by multiple submitters</li>
</ul></li>
<li><code>GDS</code>: DataSet
<ul>
<li><strong><em>Profiles</em></strong> are derived from DataSets
<ul>
<li>A Profile consists of the expression measurements for an individual gene across all Samples in a DataSet</li>
</ul></li>
</ul></li>
</ul>
<p>To establish a connection with the SQLite database:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'># Establish connection to GEOmetadb.sqlite</span>
<span class='va'>con</span> <span class='op'>&lt;-</span> <span class='fu'>DBI</span><span class='fu'>::</span><span class='fu'><a href='https://dbi.r-dbi.org/reference/dbConnect.html'>dbConnect</a></span><span class='op'>(</span>drv    <span class='op'>=</span> <span class='fu'>RSQLite</span><span class='fu'>::</span><span class='fu'><a href='https://rsqlite.r-dbi.org/reference/SQLite.html'>SQLite</a></span><span class='op'>(</span><span class='op'>)</span>,   <span class='co'>## an object of SQLiteDriver</span>
                      dbname <span class='op'>=</span> <span class='va'>metadb</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Note that packages <code>DBI</code> and <code>RSQLite</code> have already been implicitly installed as dependencies for <code>GEOmetadb</code>. Therefore we can directly make the function calls here.</p>
<p>To submit the query to the database:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>result</span> <span class='op'>&lt;-</span> <span class='fu'>DBI</span><span class='fu'>::</span><span class='fu'><a href='https://dbi.r-dbi.org/reference/dbGetQuery.html'>dbGetQuery</a></span><span class='op'>(</span>conn <span class='op'>=</span> <span class='va'>con</span>, statement <span class='op'>=</span> <span class='va'>query</span><span class='op'>)</span> <span class='co'>## save query result</span>
</code></pre>
</div>
</div>
<p>Closing connection to the database:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>DBI</span><span class='fu'>::</span><span class='fu'><a href='https://dbi.r-dbi.org/reference/dbDisconnect.html'>dbDisconnect</a></span><span class='op'>(</span><span class='va'>con</span><span class='op'>)</span> <span class='co'>## close connection</span>
<span class='fu'><a href='https://rdrr.io/r/base/rm.html'>rm</a></span><span class='op'>(</span><span class='va'>con</span><span class='op'>)</span> <span class='co'>## not a necessary step, just wanna keep my workspace clean</span>
</code></pre>
</div>
</div>
<h3 id="filtering-query-result">Filtering Query Result</h3>
<p>The SQL query result is returned as a list object by the wrapper function in R. The result should have the following columns:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>result</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>[1] &quot;title&quot;              &quot;gse&quot;                &quot;platform&quot;          
[4] &quot;submission_date&quot;    &quot;supplementary_file&quot;</code></pre>
</div>
<p>We would like to select from only those experiments whose datasets contain RNA-seq count matrices:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>hasCounts</span> <span class='op'>&lt;-</span> <span class='va'>result</span><span class='op'>$</span><span class='va'>supplementary_file</span><span class='op'>[</span>
    <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grep</a></span><span class='op'>(</span>
        <span class='va'>result</span><span class='op'>$</span><span class='va'>supplementary_file</span>,
        pattern     <span class='op'>=</span> <span class='st'>"count"</span>, <span class='co'>## keyword to look for in experiment data files</span>
        ignore.case <span class='op'>=</span> <span class='cn'>TRUE</span>
        <span class='op'>)</span>
    <span class='op'>]</span>
</code></pre>
</div>
</div>
<p>This extracts a vector of FTP links to the files that include RNA-seq counts. Then from the FTP links, we extract the GSE series numbers from their file names using regex:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>hasCounts_gse</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span>
    <span class='fu'><a href='https://rdrr.io/r/base/regmatches.html'>regmatches</a></span><span class='op'>(</span><span class='va'>hasCounts</span>,
               <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>regexec</a></span><span class='op'>(</span><span class='st'>"GSE[0-9]{4,}[^/]"</span>, <span class='va'>hasCounts</span><span class='op'>)</span>
    <span class='op'>)</span>
<span class='op'>)</span>
<span class='va'>SELECT_ROWS</span>       <span class='op'>&lt;-</span> <span class='va'>result</span><span class='op'>$</span><span class='va'>gse</span> <span class='op'>%in%</span> <span class='va'>hasCounts_gse</span>
<span class='va'>candidate_dataset</span> <span class='op'>&lt;-</span> <span class='va'>result</span><span class='op'>[</span><span class='va'>SELECT_ROWS</span>, <span class='op'>]</span>
</code></pre>
</div>
</div>
<p>The GSE series numbers and titles of the qualified candidate experiments (<code>candidate_dataset</code>) that contain RNA-seq data are as follows:</p>
<div class="layout-chunk" data-layout="l-body">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:200px; overflow-x: scroll; width:100%; ">
<table class=" lightable-paper" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
title
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
gse
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
Transcriptional response of human lung epithelial cells to SARS-CoV-2 infection
</td>
<td style="text-align:left;">
GSE147507
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
SARS-CoV-2 receptor ACE2 is an interferon-stimulated gene in human airway epithelial cells and is detected in specific cell subsets across tissues
</td>
<td style="text-align:left;">
GSE148829
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
SARS-CoV-2 receptor ACE2 is an interferon-stimulated gene in human airway epithelial cells and is detected in specific cell subsets across tissues
</td>
<td style="text-align:left;">
GSE148829
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
Spectrum of Viral Load and Host Response Seen in Autopsies of SARS-CoV-2 Infected Lungs
</td>
<td style="text-align:left;">
GSE150316
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
RNA-Seq of Human iPSC-cardiomyocytes infected with SARS-CoV-2
</td>
<td style="text-align:left;">
GSE150392
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
In vivo antiviral host response to SARS-CoV-2 by viral load, sex, and age
</td>
<td style="text-align:left;">
GSE152075
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
Severely ill COVID-19 patients display a defective exhaustion program in SARS-CoV-2 reactive CD8+ T cells
</td>
<td style="text-align:left;">
GSE153931
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:left;">
Primary Human Airway Epithelial Cultures infected with SARS-CoV-2
</td>
<td style="text-align:left;">
GSE153970
</td>
</tr>
<tr>
<td style="text-align:left;">
14
</td>
<td style="text-align:left;">
In vivo antiviral host response to SARS-CoV-2 by viral load, sex, and age [dataset II]
</td>
<td style="text-align:left;">
GSE154768
</td>
</tr>
<tr>
<td style="text-align:left;">
15
</td>
<td style="text-align:left;">
In vivo antiviral host response to SARS-CoV-2 by viral load, sex, and age [dataset III]
</td>
<td style="text-align:left;">
GSE154769
</td>
</tr>
<tr>
<td style="text-align:left;">
22
</td>
<td style="text-align:left;">
Upper airway gene expression differentiates COVID-19 from other acute respiratory illnesses and reveals suppression of innate immune responses by SARS-CoV-2
</td>
<td style="text-align:left;">
GSE156063
</td>
</tr>
<tr>
<td style="text-align:left;">
24
</td>
<td style="text-align:left;">
Acute SARS-CoV-2 Infection in a Universal In Vivo Model for Potential Human Pandemic Coronaviruses is Cytopathic and Elicits a Robust and Sustained Host Innate Immune Response
</td>
<td style="text-align:left;">
GSE155286
</td>
</tr>
<tr>
<td style="text-align:left;">
25
</td>
<td style="text-align:left;">
Human Pluripotent Stem Cell-Derived Neural Cells and Brain Organoids Reveal SARS-CoV-2 Neurotropism Predominates in Choroid Plexus Epithelium
</td>
<td style="text-align:left;">
GSE157852
</td>
</tr>
<tr>
<td style="text-align:left;">
28
</td>
<td style="text-align:left;">
Expression of SARS-CoV-2 receptor ACE2 and coincident host response signature varies by asthma inflammatory phenotype
</td>
<td style="text-align:left;">
GSE158752
</td>
</tr>
<tr>
<td style="text-align:left;">
30
</td>
<td style="text-align:left;">
SARS-CoV-2 infected cells present HLA-I peptides from canonical and out-of-frame ORFs
</td>
<td style="text-align:left;">
GSE159191
</td>
</tr>
<tr>
<td style="text-align:left;">
31
</td>
<td style="text-align:left;">
SARS-CoV-2 infected cells present HLA-I peptides from canonical and out-of-frame ORFs
</td>
<td style="text-align:left;">
GSE159191
</td>
</tr>
<tr>
<td style="text-align:left;">
34
</td>
<td style="text-align:left;">
Cardiomyocytes Infected by SARS-CoV-2 Recruit Cardiotoxic Macrophages [bulk RNA-seq dataset)
</td>
<td style="text-align:left;">
GSE151879
</td>
</tr>
<tr>
<td style="text-align:left;">
35
</td>
<td style="text-align:left;">
Cardiomyocytes Infected by SARS-CoV-2 Recruit Cardiotoxic Macrophages [bulk RNA-seq dataset)
</td>
<td style="text-align:left;">
GSE151879
</td>
</tr>
<tr>
<td style="text-align:left;">
42
</td>
<td style="text-align:left;">
Identification of required host factors for SARS-CoV-2 infection in human cells [RNA-seq]
</td>
<td style="text-align:left;">
GSE159522
</td>
</tr>
<tr>
<td style="text-align:left;">
44
</td>
<td style="text-align:left;">
SARS-CoV-2 infection of primary human lung epithelium for COVID-19 modeling and drug discovery
</td>
<td style="text-align:left;">
GSE155518
</td>
</tr>
<tr>
<td style="text-align:left;">
48
</td>
<td style="text-align:left;">
Temporal and Spatial Heterogeneity of Host Response to SARS-CoV-2 Pulmonary Infection [gene expression]
</td>
<td style="text-align:left;">
GSE159787
</td>
</tr>
<tr>
<td style="text-align:left;">
53
</td>
<td style="text-align:left;">
Single Nucleus RNA-Seq Multiomic Profiling Reveals Age-Dynamic Regulation of Host Genes Associated with SARS-CoV-2 Infection [scRNA-Seq]
</td>
<td style="text-align:left;">
GSE161382
</td>
</tr>
<tr>
<td style="text-align:left;">
56
</td>
<td style="text-align:left;">
Genome-scale identification of SARS-CoV-2 and pan-coronavirus host factor networks
</td>
<td style="text-align:left;">
GSE162038
</td>
</tr>
<tr>
<td style="text-align:left;">
57
</td>
<td style="text-align:left;">
Functional interrogation of a SARS-CoV-2 host protein interactome identifies unique and shared coronavirus host factors
</td>
<td style="text-align:left;">
GSE162039
</td>
</tr>
<tr>
<td style="text-align:left;">
58
</td>
<td style="text-align:left;">
Metabolic reprogramming and epigenetic changes of vital organs in SARS-CoV-2 induced systemic toxicity
</td>
<td style="text-align:left;">
GSE162113
</td>
</tr>
<tr>
<td style="text-align:left;">
59
</td>
<td style="text-align:left;">
Infection of human Nasal Epithelial Cells with SARS-CoV-2 and a 382-nt deletion isolate lacking ORF8 reveals similar viral kinetics and host transcriptional profiles
</td>
<td style="text-align:left;">
GSE162131
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: white !important;background-color: #D7261E !important;">
63
</td>
<td style="text-align:left;font-weight: bold;color: white !important;background-color: #D7261E !important;">
Transcriptomic Similarities and Differences in Host Response between SARS-CoV-2 and Other Viral Infection
</td>
<td style="text-align:left;font-weight: bold;color: white !important;background-color: #D7261E !important;">
GSE152641
</td>
</tr>
<tr>
<td style="text-align:left;">
64
</td>
<td style="text-align:left;">
Bulk RNA sequencing of SARS-CoV-2 infected alveolar type I- and type II like cells
</td>
<td style="text-align:left;">
GSE153218
</td>
</tr>
<tr>
<td style="text-align:left;">
73
</td>
<td style="text-align:left;">
Identification of Drugs Blocking SARS-CoV-2 Infection using Human Pluripotent Stem Cell-derived Colonic Organoids
</td>
<td style="text-align:left;">
GSE148696
</td>
</tr>
<tr>
<td style="text-align:left;">
74
</td>
<td style="text-align:left;">
Response to SARS-CoV-2 infection in cornea, limbus and sclera from human donors
</td>
<td style="text-align:left;">
GSE164073
</td>
</tr>
<tr>
<td style="text-align:left;">
76
</td>
<td style="text-align:left;">
Dysregulated transcriptional responses to SARS-CoV-2 in the periphery support novel diagnostic approaches
</td>
<td style="text-align:left;">
GSE161731
</td>
</tr>
<tr>
<td style="text-align:left;">
78
</td>
<td style="text-align:left;">
A Multiplexed, Next-Generation Sequencing Platform for High-Throughput Detection of SARS-CoV-2 [PoC cohort]
</td>
<td style="text-align:left;">
GSE160031
</td>
</tr>
<tr>
<td style="text-align:left;">
79
</td>
<td style="text-align:left;">
A Multiplexed, Next-Generation Sequencing Platform for High-Throughput Detection of SARS-CoV-2 [test cohort]
</td>
<td style="text-align:left;">
GSE160032
</td>
</tr>
<tr>
<td style="text-align:left;">
80
</td>
<td style="text-align:left;">
A Multiplexed, Next-Generation Sequencing Platform for High-Throughput Detection of SARS-CoV-2 [pilot cohort]
</td>
<td style="text-align:left;">
GSE160033
</td>
</tr>
<tr>
<td style="text-align:left;">
81
</td>
<td style="text-align:left;">
A Multiplexed, Next-Generation Sequencing Platform for High-Throughput Detection of SARS-CoV-2 [extended cohort]
</td>
<td style="text-align:left;">
GSE160034
</td>
</tr>
<tr>
<td style="text-align:left;">
84
</td>
<td style="text-align:left;">
SARS-CoV-2 infection of human pancreatic islets
</td>
<td style="text-align:left;">
GSE159717
</td>
</tr>
<tr>
<td style="text-align:left;">
94
</td>
<td style="text-align:left;">
Expression profiling of SARS-CoV-2 infected patients by RNA-seq
</td>
<td style="text-align:left;">
GSE154244
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We choose <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152641"><strong>GSE152641</strong></a> from the candidate studies to be our experiment of interest for the following demonstration.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>series</span> <span class='op'>&lt;-</span> <span class='st'>"GSE152641"</span> <span class='co'>## GEO series ID</span>
</code></pre>
</div>
</div>
<p>Always read the publication of the original experiment and understand the experiment before starting the analysis.</p>
<h3 id="overview-of-the-chosen-experiment">Overview of the chosen experiment</h3>
<p>The original study of <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152641">GSE152641</a><span class="citation" data-cites="GSE152641">(<a href="#ref-GSE152641" role="doc-biblioref">Thair et al. 2021</a>)</span> performed expression analysis on genes of these 86 samples and compiled expression data of prior studies on six viruses: influenza, RSV, HRV, Ebola, Dengue, and SARS in an attempt to isolate COVID-19 biomarkers from other viral infections. The RNA-seq data was generated by profiling peripheral blood from <strong>62 COVID-19 patients</strong> and <strong>24 healthy controls</strong> via bulk RNA-seq. The RNA-seq count matrix in the data file consists of <strong>20460 unique Entrez gene IDs</strong> along the rows and 86 samples (biological replicates) along the columns.</p>
<h3 id="downloading-dataset">Downloading Dataset</h3>
<p><code>GEOquery</code> provides easy and quick methods for accessing data from GEO <span class="citation" data-cites="GEOquery">(<a href="#ref-GEOquery" role="doc-biblioref">Davis and Meltzer 2007</a>)</span>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"GEOquery"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>
    <span class='fu'>BiocManager</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocManager/man/install.html'>install</a></span><span class='op'>(</span><span class='st'>"GEOquery"</span>, ask <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>To download the RNA-seq data file for the corresponding GSE series number from GEO:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'>## Get names of data files</span>
<span class='va'>fname</span> <span class='op'>&lt;-</span> <span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/getGEOSuppFiles.html'>getGEOSuppFiles</a></span><span class='op'>(</span>GEO           <span class='op'>=</span> <span class='va'>series</span>,
                                   fetch_files   <span class='op'>=</span> <span class='cn'>FALSE</span>,  <span class='co'>## Don't download yet</span>
                                   makeDirectory <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span><span class='op'>$</span><span class='va'>fname</span>

<span class='co'>## Download the dataset if it's never downloaded into the data directory</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/files.html'>file.exists</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span><span class='op'>(</span><span class='op'>)</span>, <span class='st'>"data"</span>, <span class='va'>fname</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/getGEOSuppFiles.html'>getGEOSuppFiles</a></span><span class='op'>(</span>GEO           <span class='op'>=</span> <span class='va'>series</span>,
                              baseDir       <span class='op'>=</span> <span class='st'>"./data"</span>,
                              makeDirectory <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</div>
<p>The dataset file has been downloaded in <code>./data</code>. Note that this experiment has only 1 supplementary file; some experiments could have more than just one. Therefore you might want to check what’s in <code>fname</code> or look up <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi">the GEO accession viewer</a> before proceeding.</p>
<h2 id="rna-seq-counts-normalisation">RNA-seq Counts Normalisation</h2>
<p>Normalising RNA-seq counts is a necessary step of initial data processing prior to the differential expression analysis. The reason is that direct comparison of raw count values does not account for the effects of factors that introduce systematic errors such as different gene lengths and GC-content between genes within a biological replicate, different sequencing depth between biological replicates, and true biological differences in RNA composition between biological replicates, which have a significant impact on downstream analysis (e.g. differential expression analysis). The goal of normalisation is to minimise the effects of systematic errors by taking into account these factors.<br />
</p>
<p>There exist various normalisation methods. They come with different assumptions and correct for different factors. Evans, C., Hardin, J., and Stoebel, D. M. have published a detailed review on several common normalisation methods (DESeq, Med, Q, RPKM and ERPKM, TMM, UQ) and discussed how to choose the appropriate normalisation method based on their assumptions. <span class="citation" data-cites="norm">(<a href="#ref-norm" role="doc-biblioref">Evans, Hardin, and Stoebel 2016</a>)</span><br />
</p>
<p>Here we use the trimmed mean of M-values (TMM) normalisation method implemented in <code>edgeR</code> <span class="citation" data-cites="edgeR1">(<a href="#ref-edgeR1" role="doc-biblioref">Mark D. Robinson, McCarthy, and Smyth 2009</a>)</span><span class="citation" data-cites="edgeR2">(<a href="#ref-edgeR2" role="doc-biblioref">McCarthy, Chen, and Smyth 2012a</a>)</span>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"edgeR"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>
    <span class='fu'>BiocManager</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocManager/man/install.html'>install</a></span><span class='op'>(</span><span class='st'>"edgeR"</span>, ask <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h3 id="tmm-normalisation">TMM Normalisation</h3>
<p>The TMM approach assumes the majority of genes are not differentially expressed. It takes into account sequencing depth, gene length, and RNA composition, making the normalised count values comparable within and between samples. Readers can consult the original publication of TMM normalisation for complete details of this method. <span class="citation" data-cites="TMM">(<a href="#ref-TMM" role="doc-biblioref">Mark D. Robinson and Oshlack 2010</a>)</span><br />
</p>
<p>The following normalisation procedure is based on the edgeR differential expression analysis protocol, step 14. i-v. <span class="citation" data-cites="DE_protocol">(<a href="#ref-DE_protocol" role="doc-biblioref">Anders et al. 2013</a>)</span></p>
<h4 id="loading-dataset">Loading dataset</h4>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>covid19_exp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.csv</a></span><span class='op'>(</span>
    file        <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='va'>fname</span><span class='op'>)</span>,
    header      <span class='op'>=</span> <span class='cn'>TRUE</span>,
    check.names <span class='op'>=</span> <span class='cn'>FALSE</span> <span class='co'>## prevent R from messing up column names</span>
<span class='op'>)</span>
</code></pre>
</div>
</div>
<p>To view the first 5 genes of the count matrix:</p>
<div class="layout-chunk" data-layout="l-body">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:100%; overflow-x: scroll; width:700px; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00001
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00002
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00003
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00004
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00005
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00006
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00007
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00008
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00009
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00010
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00011
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00012
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00013
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00014
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00015
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00016
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00017
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00018
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00019
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00020
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00021
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00022
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00023
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00024
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00025
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00026
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00027
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00028
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00029
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00030
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00031
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00032
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00033
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00034
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00035
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00036
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00037
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00038
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00039
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00040
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00041
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00042
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00043
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00044
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00045
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00046
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00047
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00048
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00049
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00050
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00051
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00052
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00053
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00054
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00055
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00056
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00057
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00058
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00059
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00060
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00061
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00062
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00063
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00064
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00065
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00066
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00067
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00068
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00069
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00070
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00071
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00072
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00073
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00074
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00075
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00076
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00077
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00078
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00079
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00080
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00081
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00082
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00083
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00084
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00085
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
IMX_sample00086
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
19
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
27
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
15
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2670
</td>
<td style="text-align:right;">
2432
</td>
<td style="text-align:right;">
781
</td>
<td style="text-align:right;">
2381
</td>
<td style="text-align:right;">
2622
</td>
<td style="text-align:right;">
3273
</td>
<td style="text-align:right;">
3969
</td>
<td style="text-align:right;">
2597
</td>
<td style="text-align:right;">
1692
</td>
<td style="text-align:right;">
2571
</td>
<td style="text-align:right;">
898
</td>
<td style="text-align:right;">
1405
</td>
<td style="text-align:right;">
1132
</td>
<td style="text-align:right;">
1255
</td>
<td style="text-align:right;">
1560
</td>
<td style="text-align:right;">
585
</td>
<td style="text-align:right;">
1075
</td>
<td style="text-align:right;">
925
</td>
<td style="text-align:right;">
2989
</td>
<td style="text-align:right;">
6413
</td>
<td style="text-align:right;">
999
</td>
<td style="text-align:right;">
3950
</td>
<td style="text-align:right;">
2192
</td>
<td style="text-align:right;">
2810
</td>
<td style="text-align:right;">
829
</td>
<td style="text-align:right;">
1461
</td>
<td style="text-align:right;">
900
</td>
<td style="text-align:right;">
1261
</td>
<td style="text-align:right;">
191
</td>
<td style="text-align:right;">
1113
</td>
<td style="text-align:right;">
57
</td>
<td style="text-align:right;">
1609
</td>
<td style="text-align:right;">
2674
</td>
<td style="text-align:right;">
242
</td>
<td style="text-align:right;">
776
</td>
<td style="text-align:right;">
996
</td>
<td style="text-align:right;">
904
</td>
<td style="text-align:right;">
837
</td>
<td style="text-align:right;">
943
</td>
<td style="text-align:right;">
543
</td>
<td style="text-align:right;">
1669
</td>
<td style="text-align:right;">
651
</td>
<td style="text-align:right;">
259
</td>
<td style="text-align:right;">
1177
</td>
<td style="text-align:right;">
954
</td>
<td style="text-align:right;">
1871
</td>
<td style="text-align:right;">
553
</td>
<td style="text-align:right;">
1368
</td>
<td style="text-align:right;">
646
</td>
<td style="text-align:right;">
1607
</td>
<td style="text-align:right;">
1021
</td>
<td style="text-align:right;">
269
</td>
<td style="text-align:right;">
1323
</td>
<td style="text-align:right;">
1039
</td>
<td style="text-align:right;">
471
</td>
<td style="text-align:right;">
738
</td>
<td style="text-align:right;">
872
</td>
<td style="text-align:right;">
2245
</td>
<td style="text-align:right;">
989
</td>
<td style="text-align:right;">
2009
</td>
<td style="text-align:right;">
591
</td>
<td style="text-align:right;">
656
</td>
<td style="text-align:right;">
107
</td>
<td style="text-align:right;">
338
</td>
<td style="text-align:right;">
1285
</td>
<td style="text-align:right;">
850
</td>
<td style="text-align:right;">
861
</td>
<td style="text-align:right;">
3560
</td>
<td style="text-align:right;">
4959
</td>
<td style="text-align:right;">
1602
</td>
<td style="text-align:right;">
1696
</td>
<td style="text-align:right;">
1328
</td>
<td style="text-align:right;">
1419
</td>
<td style="text-align:right;">
475
</td>
<td style="text-align:right;">
369
</td>
<td style="text-align:right;">
736
</td>
<td style="text-align:right;">
351
</td>
<td style="text-align:right;">
1327
</td>
<td style="text-align:right;">
1458
</td>
<td style="text-align:right;">
557
</td>
<td style="text-align:right;">
591
</td>
<td style="text-align:right;">
824
</td>
<td style="text-align:right;">
643
</td>
<td style="text-align:right;">
1403
</td>
<td style="text-align:right;">
393
</td>
<td style="text-align:right;">
1009
</td>
</tr>
<tr>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
77
</td>
<td style="text-align:right;">
129
</td>
<td style="text-align:right;">
87
</td>
<td style="text-align:right;">
98
</td>
<td style="text-align:right;">
81
</td>
<td style="text-align:right;">
86
</td>
<td style="text-align:right;">
124
</td>
<td style="text-align:right;">
74
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
82
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
54
</td>
<td style="text-align:right;">
66
</td>
<td style="text-align:right;">
50
</td>
<td style="text-align:right;">
31
</td>
<td style="text-align:right;">
81
</td>
<td style="text-align:right;">
112
</td>
<td style="text-align:right;">
347
</td>
<td style="text-align:right;">
192
</td>
<td style="text-align:right;">
192
</td>
<td style="text-align:right;">
135
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
122
</td>
<td style="text-align:right;">
153
</td>
<td style="text-align:right;">
194
</td>
<td style="text-align:right;">
129
</td>
<td style="text-align:right;">
130
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
273
</td>
<td style="text-align:right;">
53
</td>
<td style="text-align:right;">
140
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
90
</td>
<td style="text-align:right;">
91
</td>
<td style="text-align:right;">
79
</td>
<td style="text-align:right;">
108
</td>
<td style="text-align:right;">
131
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
91
</td>
<td style="text-align:right;">
54
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
62
</td>
<td style="text-align:right;">
101
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
137
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
88
</td>
<td style="text-align:right;">
99
</td>
<td style="text-align:right;">
97
</td>
<td style="text-align:right;">
144
</td>
<td style="text-align:right;">
101
</td>
<td style="text-align:right;">
95
</td>
<td style="text-align:right;">
112
</td>
<td style="text-align:right;">
107
</td>
<td style="text-align:right;">
71
</td>
<td style="text-align:right;">
72
</td>
<td style="text-align:right;">
76
</td>
<td style="text-align:right;">
125
</td>
<td style="text-align:right;">
81
</td>
<td style="text-align:right;">
148
</td>
<td style="text-align:right;">
125
</td>
<td style="text-align:right;">
98
</td>
<td style="text-align:right;">
113
</td>
<td style="text-align:right;">
111
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
101
</td>
<td style="text-align:right;">
76
</td>
<td style="text-align:right;">
122
</td>
<td style="text-align:right;">
140
</td>
<td style="text-align:right;">
84
</td>
<td style="text-align:right;">
121
</td>
<td style="text-align:right;">
106
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
134
</td>
<td style="text-align:right;">
90
</td>
<td style="text-align:right;">
141
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
82
</td>
<td style="text-align:right;">
91
</td>
<td style="text-align:right;">
48
</td>
<td style="text-align:right;">
25
</td>
</tr>
<tr>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that the name for the column of gene IDs is blank. Label the column to avoid confusion:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>covid19_exp</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"entrezgene_id"</span> <span class='co'>## genes were labeled w/ Entrez IDs</span>
</code></pre>
</div>
</div>
<h4 id="defining-group">Defining group</h4>
<p>This is an experiment with <strong><em>a simple design</em></strong>: a simple comparison between two sample groups: COVID-19 and healthy control.<span class="citation" data-cites="GSE152641">(<a href="#ref-GSE152641" role="doc-biblioref">Thair et al. 2021</a>)</span><br />
</p>
<p>For most experiments in the GEO database, groups to which samples belong are usually indicated in samples’ titles. This means we are able to define groups for samples using regex to check their titles, which are the column names of the count matrix.<br />
</p>
<p>It used to be the case for GSE152641. Samples that are in the healthy control group had <code>BRH</code> or <code>HMN</code> prefix in their titles, and those in the COVID-19 group had <code>ESC</code> prefix. However, the author of GSE152641 recently replaced the prefix in all of the samples’ title with <code>IMX_sample</code>; in fact they changed the whole samples’ titles. This means we can no longer identify samples’ groups from their titles. The only way to find out their groups is look up their GSM accession numbers. Indeed, checking them one at a time on the GEO accession viewer and manually grouping them are lame.<br />
</p>
<p>We can again turn to <code>GEOquery</code> for retrieving samples’ metadata associated with their GSM accession numbers:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>gse</span> <span class='op'>&lt;-</span> <span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/getGEO.html'>getGEO</a></span><span class='op'>(</span>GEO <span class='op'>=</span> <span class='va'>series</span>, GSEMatrix <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>gsm</span> <span class='op'>&lt;-</span> <span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/GEOData-accessors.html'>GSMList</a></span><span class='op'>(</span><span class='va'>gse</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>The metadata of the experiment GSE152641 in the GEO database is in <a href="https://www.ncbi.nlm.nih.gov/geo/info/soft.html">the SOFT format</a>. <code>GEOquery</code> can parse the metadata of GSE152641 into <code>GSE</code> class, and each sample’s GSM data is in the list under the <code>GSE</code> object.<br />
</p>
<p>To view the first 5 samples’ GSM accession numbers:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>gsm</span><span class='op'>)</span>, n <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>
</code></pre>
</div>
<pre><code>[1] &quot;GSM4622627&quot; &quot;GSM4622628&quot; &quot;GSM4622629&quot; &quot;GSM4622630&quot; &quot;GSM4622631&quot;</code></pre>
</div>
<p>And the metadata that we need from the first sample:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/GEOData-accessors.html'>Meta</a></span><span class='op'>(</span><span class='va'>gsm</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>31</span><span class='op'>)</span><span class='op'>]</span>
</code></pre>
</div>
<pre><code>$characteristics_ch1
[1] &quot;disease: Healthy control&quot; &quot;tissue: Whole blood&quot;     
[3] &quot;age: 46&quot;                  &quot;Sex: Male&quot;               

$title
[1] &quot;IMX_sample00001&quot;</code></pre>
</div>
<p>Knowing these two pieces of information for each sample of the experiment, we are now able to define groups for the samples in R. The most R-ish way I’ve come up with so far without using a for-loop:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>samples</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>gsm</span>,
                  <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='op'>{</span>
                      <span class='va'>group</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span><span class='op'>(</span>
                          <span class='fu'><a href='https://rdrr.io/r/base/regmatches.html'>regmatches</a></span><span class='op'>(</span><span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/GEOData-accessors.html'>Meta</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>$</span><span class='va'>characteristics_ch1</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,
                                     <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>regexec</a></span><span class='op'>(</span>
                                         pattern <span class='op'>=</span> <span class='st'>"(\\s).*"</span>,  <span class='co'>## match by space</span>
                                         <span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/GEOData-accessors.html'>Meta</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>$</span><span class='va'>characteristics_ch1</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span>
                                     <span class='op'>)</span>
                          <span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span> <span class='co'>## using regex to extract sample group</span>
                      <span class='co'>## remove space at the beginning of a string</span>
                      <span class='va'>group</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/grep.html'>gsub</a></span><span class='op'>(</span>pattern <span class='op'>=</span> <span class='st'>"^\\s"</span>, <span class='st'>""</span>, <span class='va'>group</span><span class='op'>)</span>
                      <span class='va'>sample</span> <span class='op'>&lt;-</span> <span class='fu'>GEOquery</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GEOquery/man/GEOData-accessors.html'>Meta</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>$</span><span class='va'>title</span>
                      <span class='kw'><a href='https://rdrr.io/r/base/function.html'>return</a></span><span class='op'>(</span>
                          <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
                              sample      <span class='op'>=</span> <span class='va'>sample</span>,
                              group       <span class='op'>=</span> <span class='va'>group</span>,
                              check.names <span class='op'>=</span> <span class='cn'>FALSE</span>
                              <span class='op'>)</span>
                          <span class='op'>)</span>
                    <span class='op'>}</span>
                  <span class='op'>)</span>
<span class='va'>samples</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span><span class='va'>rbind</span>, <span class='va'>samples</span><span class='op'>)</span> <span class='co'>## combine the elements of a list</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:200px; overflow-x: scroll; width:100%; ">
<table class=" lightable-paper" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
sample
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
group
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
GSM4622627
</td>
<td style="text-align:left;">
IMX_sample00001
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622628
</td>
<td style="text-align:left;">
IMX_sample00002
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622629
</td>
<td style="text-align:left;">
IMX_sample00003
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622630
</td>
<td style="text-align:left;">
IMX_sample00004
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622631
</td>
<td style="text-align:left;">
IMX_sample00005
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622632
</td>
<td style="text-align:left;">
IMX_sample00006
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622633
</td>
<td style="text-align:left;">
IMX_sample00007
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622634
</td>
<td style="text-align:left;">
IMX_sample00008
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622635
</td>
<td style="text-align:left;">
IMX_sample00009
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622636
</td>
<td style="text-align:left;">
IMX_sample00010
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622637
</td>
<td style="text-align:left;">
IMX_sample00011
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622638
</td>
<td style="text-align:left;">
IMX_sample00012
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622639
</td>
<td style="text-align:left;">
IMX_sample00013
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622640
</td>
<td style="text-align:left;">
IMX_sample00014
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622641
</td>
<td style="text-align:left;">
IMX_sample00015
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622642
</td>
<td style="text-align:left;">
IMX_sample00016
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622643
</td>
<td style="text-align:left;">
IMX_sample00017
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622644
</td>
<td style="text-align:left;">
IMX_sample00018
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622645
</td>
<td style="text-align:left;">
IMX_sample00019
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622646
</td>
<td style="text-align:left;">
IMX_sample00020
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622647
</td>
<td style="text-align:left;">
IMX_sample00021
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622648
</td>
<td style="text-align:left;">
IMX_sample00022
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622649
</td>
<td style="text-align:left;">
IMX_sample00023
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622650
</td>
<td style="text-align:left;">
IMX_sample00024
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622651
</td>
<td style="text-align:left;">
IMX_sample00025
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622652
</td>
<td style="text-align:left;">
IMX_sample00026
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622653
</td>
<td style="text-align:left;">
IMX_sample00027
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622654
</td>
<td style="text-align:left;">
IMX_sample00028
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622655
</td>
<td style="text-align:left;">
IMX_sample00029
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622656
</td>
<td style="text-align:left;">
IMX_sample00030
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622657
</td>
<td style="text-align:left;">
IMX_sample00031
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622658
</td>
<td style="text-align:left;">
IMX_sample00032
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622659
</td>
<td style="text-align:left;">
IMX_sample00033
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622660
</td>
<td style="text-align:left;">
IMX_sample00034
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622661
</td>
<td style="text-align:left;">
IMX_sample00035
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622662
</td>
<td style="text-align:left;">
IMX_sample00036
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622663
</td>
<td style="text-align:left;">
IMX_sample00037
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622664
</td>
<td style="text-align:left;">
IMX_sample00038
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622665
</td>
<td style="text-align:left;">
IMX_sample00039
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622666
</td>
<td style="text-align:left;">
IMX_sample00040
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622667
</td>
<td style="text-align:left;">
IMX_sample00041
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622668
</td>
<td style="text-align:left;">
IMX_sample00042
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622669
</td>
<td style="text-align:left;">
IMX_sample00043
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622670
</td>
<td style="text-align:left;">
IMX_sample00044
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622671
</td>
<td style="text-align:left;">
IMX_sample00045
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622672
</td>
<td style="text-align:left;">
IMX_sample00046
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622673
</td>
<td style="text-align:left;">
IMX_sample00047
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622674
</td>
<td style="text-align:left;">
IMX_sample00048
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622675
</td>
<td style="text-align:left;">
IMX_sample00049
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622676
</td>
<td style="text-align:left;">
IMX_sample00050
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622677
</td>
<td style="text-align:left;">
IMX_sample00051
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622678
</td>
<td style="text-align:left;">
IMX_sample00052
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622679
</td>
<td style="text-align:left;">
IMX_sample00053
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622680
</td>
<td style="text-align:left;">
IMX_sample00054
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622681
</td>
<td style="text-align:left;">
IMX_sample00055
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622682
</td>
<td style="text-align:left;">
IMX_sample00056
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622683
</td>
<td style="text-align:left;">
IMX_sample00057
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622684
</td>
<td style="text-align:left;">
IMX_sample00058
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622685
</td>
<td style="text-align:left;">
IMX_sample00059
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622686
</td>
<td style="text-align:left;">
IMX_sample00060
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622687
</td>
<td style="text-align:left;">
IMX_sample00061
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622688
</td>
<td style="text-align:left;">
IMX_sample00062
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622689
</td>
<td style="text-align:left;">
IMX_sample00063
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622690
</td>
<td style="text-align:left;">
IMX_sample00064
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622691
</td>
<td style="text-align:left;">
IMX_sample00065
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622692
</td>
<td style="text-align:left;">
IMX_sample00066
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622693
</td>
<td style="text-align:left;">
IMX_sample00067
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622694
</td>
<td style="text-align:left;">
IMX_sample00068
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622695
</td>
<td style="text-align:left;">
IMX_sample00069
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622696
</td>
<td style="text-align:left;">
IMX_sample00070
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622697
</td>
<td style="text-align:left;">
IMX_sample00071
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622698
</td>
<td style="text-align:left;">
IMX_sample00072
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622699
</td>
<td style="text-align:left;">
IMX_sample00073
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622700
</td>
<td style="text-align:left;">
IMX_sample00074
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622701
</td>
<td style="text-align:left;">
IMX_sample00075
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622702
</td>
<td style="text-align:left;">
IMX_sample00076
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622703
</td>
<td style="text-align:left;">
IMX_sample00077
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622704
</td>
<td style="text-align:left;">
IMX_sample00078
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622705
</td>
<td style="text-align:left;">
IMX_sample00079
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622706
</td>
<td style="text-align:left;">
IMX_sample00080
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622707
</td>
<td style="text-align:left;">
IMX_sample00081
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622708
</td>
<td style="text-align:left;">
IMX_sample00082
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622709
</td>
<td style="text-align:left;">
IMX_sample00083
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622710
</td>
<td style="text-align:left;">
IMX_sample00084
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622711
</td>
<td style="text-align:left;">
IMX_sample00085
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622712
</td>
<td style="text-align:left;">
IMX_sample00086
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Save this data frame as an R object. We will need to use it later when performing differential gene expression analysis:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>samples</span>, file <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"samples.rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h4 id="normalising-counts">Normalising counts</h4>
<p>Create a numeric matrix of the read counts with the gene IDs as row names:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>counts</span>           <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>covid19_exp</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>:</span><span class='fl'>87</span><span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>counts</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>covid19_exp</span><span class='op'>$</span><span class='va'>entrezgene_id</span>
</code></pre>
</div>
</div>
<p>According to the edgeR differential expression analysis protocol<span class="citation" data-cites="DE_protocol">(<a href="#ref-DE_protocol" role="doc-biblioref">Anders et al. 2013</a>)</span>:</p>
<blockquote>
<p>“it is recommended to remove features without at least 1 read per million in n of the samples, where n is the size of the smallest group of replicates”</p>
</blockquote>
<p>The reason is that a gene must be expressed at some minimal level before it can be translated into a protein. This means genes with very low counts across all libraries provide little evidence for differential expression.</p>
<p>Since we have only two groups: 24 healthy human controls and 62 COVID-19 patients, 24 is the size is the smallest group of biological replicates here:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>cpms</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='va'>counts</span><span class='op'>)</span>
<span class='va'>keep</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='va'>cpms</span> <span class='op'>&gt;</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>&gt;=</span> <span class='fl'>24</span>
</code></pre>
</div>
</div>
<p>Or alternatively, by the Bioconductor edgeR user guide:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>keep</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/filterByExpr.html'>filterByExpr</a></span><span class='op'>(</span><span class='va'>cpms</span>, group <span class='op'>=</span> <span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>It ultimately comes down to users’ preferences for whichever method to use; they produce the same result.<br />
</p>
<p>And we only keep genes that meet the threshold of at least 1 read per million in 24 of the samples:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>counts</span> <span class='op'>&lt;-</span> <span class='va'>counts</span><span class='op'>[</span><span class='va'>keep</span>, <span class='op'>]</span>
</code></pre>
</div>
</div>
<p>edgeR stores data in its own data class <code>DGEList</code> and performs analyses on it. Hence we need to first create a <code>DGEList</code> to store the read counts.<br />
</p>
<p>Note that although the edgeR documentation says the <code>DGEList</code> constructor takes a numeric matrix for its <code>counts</code> argument, the edgeR user guide says the <code>counts</code> argument can take either a matrix or a <code>data.frame</code> object.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>d</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>DGEList</a></span><span class='op'>(</span>
  counts <span class='op'>=</span> <span class='va'>counts</span>,
  group  <span class='op'>=</span> <span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span> <span class='co'>## add the grouping factor</span>
  <span class='op'>)</span>
</code></pre>
</div>
</div>
<p>To estimate the scale factors between samples for TMM normalization:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>d</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/calcNormFactors.html'>calcNormFactors</a></span><span class='op'>(</span><span class='va'>d</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Extract the normalised count matrix:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>normalised_counts</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/cpm.html'>cpm</a></span><span class='op'>(</span><span class='va'>d</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Save the normalised count matrix and the DGEList as R objects for later analyses:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>d</span>, file <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"DGEList.rds"</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span>object <span class='op'>=</span> <span class='va'>normalised_counts</span>,
        file   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"normalised_counts.rds"</span><span class='op'>)</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h3 id="distribution-for-read-counts---before-vs.-after-normalisation">Distribution for read counts - Before vs. after normalisation</h3>
<p>Even though we are comparing the distributions for pre-normalised read counts and post-normalised read counts, we still normalise the pre-normalised read counts to counts per million (CPM) to account for differences in library sizes between samples for a comparable result. More precisely, we are comparing the distributions for two-fold changes in expression before and after TMM normalisation. Note that genes that were previously removed according to the protocol are also excluded in this comparison.</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/boxplot.html'>boxplot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>counts</span><span class='op'>)</span>,
        xlab     <span class='op'>=</span> <span class='st'>"Samples"</span>,
        ylab     <span class='op'>=</span> <span class='st'>"log2 CPM"</span>,
        las      <span class='op'>=</span> <span class='fl'>2</span>,
        cex      <span class='op'>=</span> <span class='fl'>0.5</span>,
        cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>,
        cex.axis <span class='op'>=</span> <span class='fl'>0.25</span>,
        main     <span class='op'>=</span> <span class='st'>"Before TMM Normalisation"</span>,
        cex.main <span class='op'>=</span> <span class='fl'>0.75</span><span class='op'>)</span>

<span class='co'>#draw the median on each box plot</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/median.html'>median</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>counts</span><span class='op'>)</span>, MARGIN <span class='op'>=</span> <span class='fl'>2</span>, FUN <span class='op'>=</span> <span class='va'>median</span><span class='op'>)</span><span class='op'>)</span>,
       col <span class='op'>=</span> <span class='st'>"green"</span>,
       lwd <span class='op'>=</span> <span class='fl'>0.75</span>,
       lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/boxplot.html'>boxplot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>,
        xlab     <span class='op'>=</span> <span class='st'>"Samples"</span>,
        ylab     <span class='op'>=</span> <span class='st'>"log2 CPM"</span>,
        las      <span class='op'>=</span> <span class='fl'>2</span>,
        cex      <span class='op'>=</span> <span class='fl'>0.5</span>,
        cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>,
        cex.axis <span class='op'>=</span> <span class='fl'>0.25</span>,
        main     <span class='op'>=</span> <span class='st'>"After TMM Normalisation"</span>,
        cex.main <span class='op'>=</span> <span class='fl'>0.75</span><span class='op'>)</span>
<span class='co'># draw the median on each box plot</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/median.html'>median</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>, MARGIN <span class='op'>=</span> <span class='fl'>2</span>,FUN <span class='op'>=</span> <span class='va'>median</span><span class='op'>)</span><span class='op'>)</span>,
       col <span class='op'>=</span> <span class='st'>"green"</span>,
       lwd <span class='op'>=</span> <span class='fl'>0.75</span>,
       lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/boxplot-1.png" width="624" /></p>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>

<span class='va'>log2CPM_density</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>counts</span><span class='op'>)</span>,
                         MARGIN <span class='op'>=</span> <span class='fl'>2</span>,
                         FUN    <span class='op'>=</span> <span class='va'>density</span><span class='op'>)</span>

<span class='co'>## Get the normalised fold change cpm</span>
<span class='va'>normalised_log2CPM_density</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>,
                                    MARGIN <span class='op'>=</span> <span class='fl'>2</span>,
                                    FUN    <span class='op'>=</span> <span class='va'>density</span><span class='op'>)</span>

<span class='co'>## Calculate the limits across all the samples</span>
<span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fl'>0</span>; <span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fl'>0</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>log2CPM_density</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>xlim</span>, <span class='va'>log2CPM_density</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>$</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>
    <span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>ylim</span>, <span class='va'>log2CPM_density</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>$</span><span class='va'>y</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>normalised_log2CPM_density</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>xlim</span>, <span class='va'>normalised_log2CPM_density</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>$</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>
    <span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>ylim</span>, <span class='va'>normalised_log2CPM_density</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>$</span><span class='va'>y</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='co'>## 86 line colours for all 86 samples</span>
<span class='va'>cols</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>rainbow</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>log2CPM_density</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'>## choose line type 1 for density graphs of all 86 samples</span>
<span class='va'>ltys</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>log2CPM_density</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'>## Initialize the density plot without density lines</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>log2CPM_density</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
     xlim     <span class='op'>=</span> <span class='va'>xlim</span>,
     ylim     <span class='op'>=</span> <span class='va'>ylim</span>,
     type     <span class='op'>=</span> <span class='st'>"n"</span>,
     ylab     <span class='op'>=</span> <span class='st'>"Smoothing density of log2-CPM"</span>,
     main     <span class='op'>=</span> <span class='st'>"Before TMM normalisation"</span>,
     cex.main <span class='op'>=</span> <span class='fl'>0.7</span>,
     cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>
     <span class='op'>)</span>
<span class='co'>## Plot smoothing density for each sample</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>log2CPM_density</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='va'>log2CPM_density</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>,
          col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>)</span>
    <span class='op'>}</span>
<span class='co'>## Create legend for samples</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topright"</span>,
       <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>counts</span><span class='op'>)</span>,
       col      <span class='op'>=</span> <span class='va'>cols</span>,
       lty      <span class='op'>=</span> <span class='va'>ltys</span>,
       cex      <span class='op'>=</span> <span class='fl'>0.15</span>,
       border   <span class='op'>=</span> <span class='st'>"blue"</span>,
       text.col <span class='op'>=</span> <span class='st'>"green4"</span>,
       merge    <span class='op'>=</span> <span class='cn'>TRUE</span>,
       bg       <span class='op'>=</span> <span class='st'>"gray90"</span><span class='op'>)</span>





<span class='co'>## Initialize the density plot without density lines</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>normalised_log2CPM_density</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>,
     xlim     <span class='op'>=</span> <span class='va'>xlim</span>,
     ylim     <span class='op'>=</span> <span class='va'>ylim</span>,
     type     <span class='op'>=</span> <span class='st'>"n"</span>,
     ylab     <span class='op'>=</span> <span class='st'>"Smoothing density of log2-CPM"</span>,
     main     <span class='op'>=</span> <span class='st'>"After TMM normalisation"</span>,
     cex.main <span class='op'>=</span> <span class='fl'>0.65</span>,
     cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>,
     <span class='op'>)</span>

<span class='co'>## Plot smoothing density for each sample</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>normalised_log2CPM_density</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='va'>normalised_log2CPM_density</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>,
          col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>)</span>
    <span class='op'>}</span>
<span class='co'>## Create legend for samples</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topright"</span>,
       <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>,
       col      <span class='op'>=</span> <span class='va'>cols</span>,
       lty      <span class='op'>=</span> <span class='va'>ltys</span>,
       cex      <span class='op'>=</span> <span class='fl'>0.15</span>,
       border   <span class='op'>=</span> <span class='st'>"blue"</span>,
       text.col <span class='op'>=</span> <span class='st'>"green4"</span>,
       merge    <span class='op'>=</span> <span class='cn'>TRUE</span>,
       bg       <span class='op'>=</span> <span class='st'>"gray90"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/density-1.png" width="624" /></p>
</div>
<p>The medians of distributions for normalised RNA-seq counts are much closer to each other. The interquartile ranges for the gene expression levels of each sample also become more consistent after normalisation applied. This means 50% of gene expression levels have similar distributions for each sample.</p>
<h2 id="post-normalisation-analysis">Post-normalisation Analysis</h2>
<h3 id="biological-coefficient-of-variation-bcv">Biological coefficient of variation (BCV)</h3>
<p>To capture the variability in gene expressions between sample groups, we use group as target to build a model matrix:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>model_design</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/model.matrix.html'>model.matrix</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>group</span>, data <span class='op'>=</span> <span class='va'>samples</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:200px; overflow-x: scroll; width:100%; ">
<table class=" lightable-paper" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
(Intercept)
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
groupHealthy control
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
GSM4622627
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622628
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622629
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622630
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622631
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622632
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622633
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622634
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622635
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622636
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622637
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622638
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622639
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622640
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622641
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622642
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622643
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622644
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622645
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622646
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622647
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622648
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622649
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622650
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622651
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622652
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622653
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622654
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622655
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622656
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622657
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622658
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622659
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622660
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622661
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622662
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622663
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622664
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622665
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622666
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622667
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622668
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622669
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622670
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622671
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622672
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622673
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622674
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622675
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622676
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622677
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622678
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622679
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622680
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622681
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622682
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622683
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622684
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622685
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622686
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622687
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622688
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622689
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622690
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622691
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622692
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622693
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622694
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622695
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622696
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622697
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622698
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622699
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622700
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622701
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622702
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622703
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622704
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622705
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622706
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622707
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622708
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622709
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622710
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622711
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622712
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>d</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/estimateDisp.html'>estimateDisp</a></span><span class='op'>(</span><span class='va'>d</span>, design <span class='op'>=</span> <span class='va'>model_design</span><span class='op'>)</span>
<span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/plotBCV.html'>plotBCV</a></span><span class='op'>(</span><span class='va'>d</span>, col.tagwise <span class='op'>=</span> <span class='st'>"black"</span>, col.common <span class='op'>=</span> <span class='st'>"red"</span>,
               main <span class='op'>=</span> <span class='st'>"Common and Tag-wise Dispersion vs Expression Levels"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-35-1.png" width="624" /></p>
</div>
<p>edgeR models gene counts with the negative binomial distribution, which assumes the variance of gene counts depends on two dispersion parameters, one of which is the negative binomial dispersion, which describes how much variance deviates from the mean, and the BCV is the square root of this dispersion parameter under the negative binomial model.</p>
<p>The BCV plot presents the estimated relative variability of expression between biological replicates; it illustrates the association between the biological CV and the average true gene abundance.</p>
<p>The common dispersion line suggests all gene expression values vary by close to a BCV value of 0.5 among replicates.<span class="citation" data-cites="bcv">(<a href="#ref-bcv" role="doc-biblioref">McCarthy, Chen, and Smyth 2012b</a>)</span> The tag-wise dispersion shows BCV values calculated individually for each gene. We observe that genes with higher true abundance (under the assumption that RNA-seq counts follow a negative binomial distribution) have lower BCV’s than genes with lower abundance. This implies that the higher the true expression level of a gene has, the lower its variation in our samples.</p>
<h3 id="mean-variance-relationship">Mean-variance relationship</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/meanvar.html'>plotMeanVar</a></span><span class='op'>(</span>
    object                       <span class='op'>=</span> <span class='va'>d</span>,
    show.raw.vars                <span class='op'>=</span> <span class='cn'>TRUE</span>,
    show.tagwise.vars            <span class='op'>=</span> <span class='cn'>TRUE</span>,
    NBline                       <span class='op'>=</span> <span class='cn'>TRUE</span>,
    show.ave.raw.vars            <span class='op'>=</span> <span class='cn'>TRUE</span>,
    show.binned.common.disp.vars <span class='op'>=</span> <span class='cn'>TRUE</span>,
    main                         <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span>
        <span class='st'>"Mean-variance Relationship for 14425 Genes of"</span>,
        <span class='st'>"COVID-19 Group and Healthy Control Group"</span><span class='op'>)</span>,
    cex.main                     <span class='op'>=</span> <span class='fl'>0.8</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-36-1.png" width="624" /></p>
</div>
<p>The mean-variance plot presents the modelling of the mean-variance relationship for the normalised expression values, which are split into bins by overall expression level in the model. The grey points represent the pooled gene-wise variances. The blue points on the mean-variance plot represent the estimated gene-wise variances. The red crosses represent the pooled gene-wise variances, while the dark red crosses represent the the average of the pooled variances for each bin of genes plotted against the average expression level of the genes in the bin. The blue line shows the mean-variance relationship for a negative binomial model with common dispersion.<span class="citation" data-cites="edgeR1">(<a href="#ref-edgeR1" role="doc-biblioref">Mark D. Robinson, McCarthy, and Smyth 2009</a>)</span><span class="citation" data-cites="edgeR2">(<a href="#ref-edgeR2" role="doc-biblioref">McCarthy, Chen, and Smyth 2012a</a>)</span> We observe that all types of variances fit well along the negative binomial line.</p>
<h3 id="sample-relations">Sample relations</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/plotMDS.DGEList.html'>plotMDS.DGEList</a></span><span class='op'>(</span>
    <span class='va'>d</span>,
    col      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#b7371d"</span>, <span class='st'>"#51af84"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span><span class='op'>]</span>,
    main     <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Variation between Samples in Expression Level"</span><span class='op'>)</span>,
    cex.main <span class='op'>=</span> <span class='fl'>1</span>,
    cex      <span class='op'>=</span> <span class='fl'>0.5</span>,
    cex.axis <span class='op'>=</span> <span class='fl'>0.75</span>,
    cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>
    <span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topleft"</span>,
       lwd      <span class='op'>=</span> <span class='fl'>2</span>,
       text.col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#51af84"</span>,<span class='st'>"#b7371d"</span><span class='op'>)</span>,
       legend   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Healthy control"</span>, <span class='st'>"COVID19"</span><span class='op'>)</span>,
       lty      <span class='op'>=</span> <span class='fl'>0</span>,
       cex      <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="geo-datasets-experiment-selection-and-initial-processing_files/figure-html5/unnamed-chunk-37-1.png" width="624" /></p>
</div>
<p>The MDS plot presents variation among samples based on the normalised gene expression. The distances between each pair of samples on the MDS plot suggest how different these samples are in terms of log fold changes of the expression levels of the filtered genes. The healthy control and the COVID-19 patient samples form separate clusters along the plot in the leading logFC dimension 1.</p>
<h2 id="to-do">To Do</h2>
<p>As the genes in this dataset were labeled with Entrez gene IDs, before performing any further analyses with the processed RNA-seq data, we need to map them to their HGNC names, which are how genes are referred in most annotation datasets and literatures, and more meaningful than simply numerical IDs. This step is supposed to be part of the initial processing; however, mappings produced by different databases vary in quality. I will compare the mapping results produced by 3 different databases and finish converting gene IDs in a later post.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-DE_protocol" class="csl-entry" role="doc-biblioentry">
Anders, Simon, Davis Mccarthy, Yunshun Chen, Michal Okoniewski, Gordon Smyth, Wolfgang Huber, and Mark Robinson. 2013. <span>“Count-Based Differential Expression Analysis of RNA Sequencing Data Using r and Bioconductor.”</span> <em>Nature Protocols</em> 8 (September): 1765–86. <a href="https://doi.org/10.1038/nprot.2013.099">https://doi.org/10.1038/nprot.2013.099</a>.
</div>
<div id="ref-GEOquery" class="csl-entry" role="doc-biblioentry">
Davis, Sean, and Paul S. Meltzer. 2007. <span>“<span class="nocase">GEOquery: a bridge between the Gene Expression Omnibus (GEO) and BioConductor</span>.”</span> <em>Bioinformatics</em> 23 (14): 1846–47. <a href="https://doi.org/10.1093/bioinformatics/btm254">https://doi.org/10.1093/bioinformatics/btm254</a>.
</div>
<div id="ref-norm" class="csl-entry" role="doc-biblioentry">
Evans, Ciaran, Johanna Hardin, and Daniel Stoebel. 2016. <span>“Selecting Between-Sample RNA-Seq Normalization Methods from the Perspective of Their Assumptions.”</span> <em>Briefings in Bioinformatics</em> 19 (September). <a href="https://doi.org/10.1093/bib/bbx008">https://doi.org/10.1093/bib/bbx008</a>.
</div>
<div id="ref-edgeR2" class="csl-entry" role="doc-biblioentry">
McCarthy, Davis J., Yunshun Chen, and Gordon K. Smyth. 2012a. <span>“<span class="nocase">Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation</span>.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97. <a href="https://doi.org/10.1093/nar/gks042">https://doi.org/10.1093/nar/gks042</a>.
</div>
<div id="ref-bcv" class="csl-entry" role="doc-biblioentry">
———. 2012b. <span>“<span class="nocase">Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation</span>.”</span> <em>Nucleic Acids Research</em> 40 (10): 4288–97. <a href="https://doi.org/10.1093/nar/gks042">https://doi.org/10.1093/nar/gks042</a>.
</div>
<div id="ref-edgeR1" class="csl-entry" role="doc-biblioentry">
Robinson, Mark D., Davis J. McCarthy, and Gordon K. Smyth. 2009. <span>“<span class="nocase">edgeR: a Bioconductor package for differential expression analysis of digital gene expression data</span>.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-TMM" class="csl-entry" role="doc-biblioentry">
Robinson, Mark D, and Alicia Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.”</span> <em>Genome Biology</em> 11 (3): 1–9.
</div>
<div id="ref-GSE152641" class="csl-entry" role="doc-biblioentry">
Thair, Simone A., Yudong D. He, Yehudit Hasin-Brumshtein, Suraj Sakaram, Rushika Pandya, Jiaying Toh, David Rawling, et al. 2021. <span>“Transcriptomic Similarities and Differences in Host Response Between SARS-CoV-2 and Other Viral Infections.”</span> <em>iScience</em> 24 (1): 101947. https://doi.org/<a href="https://doi.org/10.1016/j.isci.2020.101947">https://doi.org/10.1016/j.isci.2020.101947</a>.
</div>
<div id="ref-GEOmetadb" class="csl-entry" role="doc-biblioentry">
Zhu, Yuelin, Sean Davis, Robert Stephens, Paul S. Meltzer, and Yidong Chen. 2008. <span>“<span class="nocase">GEOmetadb: powerful alternative search engine for the Gene Expression Omnibus</span>.”</span> <em>Bioinformatics</em> 24 (23): 2798–2800. <a href="https://doi.org/10.1093/bioinformatics/btn520">https://doi.org/10.1093/bioinformatics/btn520</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
