---
title: "Comparing HGNC Symbol Mappings by 3 Different Databases"
description: |
    Mapping from Entrez gene ID's to HGNC gene symbols for genes in GSE152641 using biomaRt, BridgeDbR, and org.Hs.eg.db.
author:
    - name: Feifei Li
      url: {}
date: 2021-07-07
bibliography: biblio.bib
output:
    distill::distill_article:
        self_contained: FALSE
        toc: TRUE
creative_commons: CC BY
categories:
    - R
    - bioconductor
    - bioinformatics
preview: biomart.png
draft: TRUE
---


```{r setup, include=FALSE}
library(kableExtra)
knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE)
```

This post is a continuation of
[the initial processing of GSE152641 RNA-seq data](https://www.ffli.dev/posts/2021-07-05-geo-datasets-experiment-selection-and-initial-processing/).
The aim is to map the identifiers of genes in GSE152641
from Entrez gene ID's to the corresponding HGNC gene names
using three different databases,
compare the results of mapping by these databases, and
choose the mapping that has the best coverage.
Note that we are not mapping HGNC symbols for all 20460 genes,
but for the remaining 14426 genes
after filtering genes with low counts in
[the normalisation procedure](https://www.ffli.dev/posts/2021-07-05-geo-datasets-experiment-selection-and-initial-processing/#tmm-normalisation).

## biomaRt



```{r}
if (!requireNamespace("biomaRt", quietly = TRUE))
    BiocManager::install("biomaRt")
```

## BridgeDbR

```{r}
if (!requireNamespace("BridgeDbR", quietly = TRUE))
    BiocManager::install("BridgeDbR", ask = FALSE)
```

## org.Hs.eg.db

```{r}
if (!requireNamespace("org.Hs.eg.db", quietly = TRUE))
    BiocManager::install("org.Hs.eg.db", ask = FALSE)
```

```{r}
# Use org.Hs.eg.db database to map identifiers
if (!requireNamespace("AnnotationDbi", quietly = TRUE))
    BiocManager::install("AnnotationDbi", ask = FALSE)
```

