<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

  <!--radix_placeholder_meta_tags-->
  <title>Differential Expression Analysis: edgeR vs limma</title>

  <meta property="description" itemprop="description" content="Perform differential expression analysis&#10;and visualise differentially expressed genes in R."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2021-07-26"/>
  <meta property="article:created" itemprop="dateCreated" content="2021-07-26"/>
  <meta name="article:author" content="Feifei Li"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="Differential Expression Analysis: edgeR vs limma"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="Perform differential expression analysis&#10;and visualise differentially expressed genes in R."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="Differential Expression Analysis: edgeR vs limma"/>
  <meta property="twitter:description" content="Perform differential expression analysis&#10;and visualise differentially expressed genes in R."/>

  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Transcriptomic similarities and differences in host response between SARS-CoV-2 and other viral infections;citation_publication_date=2021;citation_volume=24;citation_doi=https://doi.org/10.1016/j.isci.2020.101947;citation_issn=2589-0042;citation_author=Simone A. Thair;citation_author=Yudong D. He;citation_author=Yehudit Hasin-Brumshtein;citation_author=Suraj Sakaram;citation_author=Rushika Pandya;citation_author=Jiaying Toh;citation_author=David Rawling;citation_author=Melissa Remmel;citation_author=Sabrina Coyle;citation_author=George N. Dalekos;citation_author=Ioannis Koutsodimitropoulos;citation_author=Glykeria Vlachogianni;citation_author=Eleni Gkeka;citation_author=Eleni Karakike;citation_author=Georgia Damoraki;citation_author=Nikolaos Antonakos;citation_author=Purvesh Khatri;citation_author=Evangelos J. Giamarellos-Bourboulis;citation_author=Timothy E. Sweeney"/>
  <meta name="citation_reference" content="citation_title=Complex heatmaps reveal patterns and correlations in multidimensional genomic data;citation_publication_date=2016;citation_volume=32;citation_doi=10.1093/bioinformatics/btw313;citation_issn=1367-4803;citation_author=Zuguang Gu;citation_author=Roland Eils;citation_author=Matthias Schlesner"/>
  <meta name="citation_reference" content="citation_title=circlize implements and enhances circular visualization in R ;citation_publication_date=2014;citation_volume=30;citation_doi=10.1093/bioinformatics/btu393;citation_issn=1367-4803;citation_author=Zuguang Gu;citation_author=Lei Gu;citation_author=Roland Eils;citation_author=Matthias Schlesner;citation_author=Benedikt Brors"/>
  <meta name="citation_reference" content="citation_title=edgeR: a Bioconductor package for differential expression analysis of digital gene expression data;citation_publication_date=2009;citation_volume=26;citation_doi=10.1093/bioinformatics/btp616;citation_issn=1367-4803;citation_author=Mark D. Robinson;citation_author=Davis J. McCarthy;citation_author=Gordon K. Smyth"/>
  <meta name="citation_reference" content="citation_title=Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation;citation_publication_date=2012;citation_volume=40;citation_doi=10.1093/nar/gks042;citation_issn=0305-1048;citation_author=Davis J. McCarthy;citation_author=Yunshun Chen;citation_author=Gordon K. Smyth"/>
  <meta name="citation_reference" content="citation_title=Interpretation of differential gene expression results of RNA-seq data: review and integration;citation_publication_date=2018;citation_volume=20;citation_doi=10.1093/bib/bby067;citation_issn=1477-4054;citation_author=Adam McDermaid;citation_author=Brandon Monier;citation_author=Jing Zhao;citation_author=Bingqiang Liu;citation_author=Qin Ma"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","bibliography","output","creative_commons","categories","preview"]}},"value":[{"type":"character","attributes":{},"value":["Differential Expression Analysis: edgeR vs limma"]},{"type":"character","attributes":{},"value":["Perform differential expression analysis\nand visualise differentially expressed genes in R.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name"]}},"value":[{"type":"character","attributes":{},"value":["Feifei Li"]}]}]},{"type":"character","attributes":{},"value":["2021-07-26"]},{"type":"character","attributes":{},"value":["biblio.bib"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["R","bioconductor","bioinformatics"]},{"type":"character","attributes":{},"value":["hm.png"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["biblio.bib","data/covid19_downregulated_genes.txt","data/covid19_ranked_genelist.txt","data/covid19_upregulated_genes.txt","data/normalised_counts.rds","data/samples.rds","differential-expression-analysis-edger-vs-limma_files/anchor-4.2.2/anchor.min.js","differential-expression-analysis-edger-vs-limma_files/bowser-1.9.3/bowser.min.js","differential-expression-analysis-edger-vs-limma_files/distill-2.2.21/template.v2.js","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-11-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-20-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-21-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-27-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-28-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-29-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-30-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-4-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-5-1.png","differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-7-1.png","differential-expression-analysis-edger-vs-limma_files/header-attrs-2.9/header-attrs.js","differential-expression-analysis-edger-vs-limma_files/jquery-1.11.3/jquery.min.js","differential-expression-analysis-edger-vs-limma_files/kePrint-0.0.1/kePrint.js","differential-expression-analysis-edger-vs-limma_files/lightable-0.0.1/lightable.css","differential-expression-analysis-edger-vs-limma_files/popper-2.6.0/popper.min.js","differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy-bundle.umd.min.js","differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy-light-border.css","differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy.css","differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy.umd.min.js","differential-expression-analysis-edger-vs-limma_files/webcomponents-2.0.0/webcomponents.js","hm.png"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          return "<p>" + $('#ref-' + ref).html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="differential-expression-analysis-edger-vs-limma_files/header-attrs-2.9/header-attrs.js"></script>
  <script src="differential-expression-analysis-edger-vs-limma_files/kePrint-0.0.1/kePrint.js"></script>
  <link href="differential-expression-analysis-edger-vs-limma_files/lightable-0.0.1/lightable.css" rel="stylesheet" />
  <script src="differential-expression-analysis-edger-vs-limma_files/jquery-1.11.3/jquery.min.js"></script>
  <script src="differential-expression-analysis-edger-vs-limma_files/popper-2.6.0/popper.min.js"></script>
  <link href="differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="differential-expression-analysis-edger-vs-limma_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="differential-expression-analysis-edger-vs-limma_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="differential-expression-analysis-edger-vs-limma_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="differential-expression-analysis-edger-vs-limma_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="differential-expression-analysis-edger-vs-limma_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Differential Expression Analysis: edgeR vs limma","description":"Perform differential expression analysis\nand visualise differentially expressed genes in R.","authors":[{"author":"Feifei Li","authorURL":"#","affiliation":"&nbsp;","affiliationURL":"#","orcidID":""}],"publishedDate":"2021-07-26T00:00:00.000-04:00","citationText":"Li, 2021"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Differential Expression Analysis: edgeR vs limma</h1>
<!--radix_placeholder_categories-->
<div class="dt-tags">
<div class="dt=tag">R</div>
<div class="dt=tag">bioconductor</div>
<div class="dt=tag">bioinformatics</div>
</div>
<!--/radix_placeholder_categories-->
<p><p>Perform differential expression analysis and visualise differentially expressed genes in R.</p></p>
</div>

<div class="d-byline">
  Feifei Li  
  
<br/>2021-07-26
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#visualising-expression-level">Visualising Expression Level</a>
<ul>
<li><a href="#standardising-expression-values">Standardising expression values</a></li>
<li><a href="#heatmap-annotation">Heatmap annotation</a></li>
</ul></li>
<li><a href="#de-analysis-by-quasi-likelihood-methods-qlm">DE Analysis by Quasi-Likelihood Methods (QLM)</a>
<ul>
<li><a href="#model-design">Model Design</a></li>
<li><a href="#multiple-hypothesis-testing-for-differentially-expressed-genes-by-qlm">Multiple Hypothesis Testing for Differentially Expressed Genes by QLM</a></li>
<li><a href="#ql-f-test-p-values-comparison-fdr-correcred-vs-uncorrected">QL F-test P-values Comparison: FDR-correcred vs Uncorrected</a></li>
<li><a href="#volcano-plot">Volcano Plot</a></li>
<li><a href="#heatmap-for-qlf-top-hit-genes">Heatmap for QLF top-hit Genes</a></li>
</ul></li>
<li><a href="#differential-expression-analysis-by-linear-models-of-mircroarray-limma">Differential Expression Analysis by Linear Models of MircroArray (Limma)</a>
<ul>
<li><a href="#multiple-hypothesis-testing">Multiple Hypothesis Testing</a></li>
<li><a href="#p-value-adjustment-methods">P-value adjustment methods</a></li>
<li><a href="#limma-p-values-comparison-bh-adjusted-vs-unadjusted">Limma P-values Comparison: BH-adjusted vs Unadjusted</a></li>
<li><a href="#volcano-plot-for-limma-top-hit-genes">Volcano Plot for Limma top-hit Genes</a></li>
<li><a href="#heatmap-for-limma-top-hit-genes">Heatmap for Limma top-hit Genes</a></li>
</ul></li>
<li><a href="#result-comparing-ql-f-test-and-limma-results">Result: Comparing QL F-test and Limma Results</a></li>
<li><a href="#to-do">To Do</a>
<ul>
<li><a href="#thresholded-list">Thresholded List</a></li>
<li><a href="#ranked-list">Ranked List</a></li>
</ul></li>
</ul>
</nav>
</div>
<p>The dataset being used here is <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152641">GSE152641</a><span class="citation" data-cites="GSE152641">(<a href="#ref-GSE152641" role="doc-biblioref">Thair et al. 2021</a>)</span> Before differential expression analysis, we have done:</p>
<ul>
<li><a href="https://www.ffli.dev/posts/2021-07-05-geo-datasets-experiment-selection-and-initial-processing/">Initial RNA-seq data processing</a></li>
<li><a href="https://www.ffli.dev/posts/2021-07-11-comparing-hgnc-symbol-mappings-by-3-different-databases/">Mapping gene identifiers to HGNC gene names</a></li>
</ul>
<p>To load the processed RNA-seq data:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>normalised_counts</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>file <span class='op'>=</span> <span class='st'>"./data/normalised_counts.rds"</span><span class='op'>)</span>
<span class='va'>samples</span>           <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span>file <span class='op'>=</span> <span class='st'>"./data/samples.rds"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h2 id="visualising-expression-level">Visualising Expression Level</h2>
<p>We will be using <a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmap</a><span class="citation" data-cites="complexheatmap">(<a href="#ref-complexheatmap" role="doc-biblioref">Gu, Eils, and Schlesner 2016</a>)</span> for heatmap visualisation of gene expression and <a href="https://github.com/jokergoo/circlize">circlize</a><span class="citation" data-cites="circlize">(<a href="#ref-circlize" role="doc-biblioref">Gu et al. 2014</a>)</span> to generate a colour gredient indicating expression levels of the genes:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"circlize"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>
    <span class='fu'><a href='https://rdrr.io/r/utils/install.packages.html'>install.packages</a></span><span class='op'>(</span><span class='st'>"circlize"</span><span class='op'>)</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"ComplexHeatmap"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span>
    <span class='fu'>BiocManager</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocManager/man/install.html'>install</a></span><span class='op'>(</span><span class='st'>"ComplexHeatmap"</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>We write a function to create a color palette using <code>circlize</code>:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>heatmap_cols</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>mat</span>, <span class='va'>neg</span> <span class='op'>=</span> <span class='st'>"blue"</span>, <span class='va'>zero</span> <span class='op'>=</span> <span class='st'>"white"</span>, <span class='va'>pos</span> <span class='op'>=</span> <span class='st'>"red"</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'># Purpose:</span>
    <span class='co'>#     Helper function to create a transition colour gradient for a heatmap</span>
    <span class='co'># Parameters:</span>
    <span class='co'>#     mat: A RNA-seq count matrix</span>
    <span class='co'>#     neg: colour for negative expression values, green by default</span>
    <span class='co'>#     zero: colour for expression values of zero, black by default</span>
    <span class='co'>#     pos: colour for positive expression values, red by default</span>
    <span class='co'># Value:</span>
    <span class='co'>#     cols: a vector of colour gradient indicating expression level</span>
    
    <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='va'>mat</span><span class='op'>)</span> <span class='op'>==</span> <span class='fl'>0</span><span class='op'>)</span> <span class='op'>{</span> <span class='co'>## when the lowest expression value is 0</span>
        <span class='va'>cols</span> <span class='op'>&lt;-</span> <span class='fu'>circlize</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/circlize/man/colorRamp2.html'>colorRamp2</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='va'>mat</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>mat</span><span class='op'>)</span><span class='op'>)</span>,
                                     <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>zero</span>, <span class='va'>pos</span><span class='op'>)</span>
                                     <span class='op'>)</span>
    <span class='op'>}</span> <span class='kw'>else</span> <span class='op'>{</span> <span class='co'>## when the lowest expression value is negative</span>
        <span class='va'>cols</span> <span class='op'>&lt;-</span> <span class='fu'>circlize</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/circlize/man/colorRamp2.html'>colorRamp2</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='va'>mat</span><span class='op'>)</span>, <span class='fl'>0</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>mat</span><span class='op'>)</span><span class='op'>)</span>,
                                     <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>neg</span>, <span class='va'>zero</span>, <span class='va'>pos</span><span class='op'>)</span>
                                     <span class='op'>)</span>
    <span class='op'>}</span>
    
    <span class='kw'><a href='https://rdrr.io/r/base/function.html'>return</a></span><span class='op'>(</span><span class='va'>cols</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<p>One might ask, how could an expression value be negative? Before answering that, if we directly plot out the gene counts in a heatmap:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>ComplexHeatmap</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html'>Heatmap</a></span><span class='op'>(</span>
    matrix              <span class='op'>=</span> <span class='va'>normalised_counts</span>,
    column_title        <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Expression Levels of"</span>,
                                <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span><span class='op'>)</span>,
                                <span class='st'>"Genes in GSE152641"</span><span class='op'>)</span>,
    name                <span class='op'>=</span> <span class='st'>"Expr lvl"</span>,
    cluster_rows        <span class='op'>=</span> <span class='cn'>TRUE</span>,
    cluster_columns     <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_row_dend       <span class='op'>=</span> <span class='cn'>TRUE</span>,
    show_column_dend    <span class='op'>=</span> <span class='cn'>TRUE</span>,
    col                 <span class='op'>=</span> <span class='fu'>heatmap_cols</span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>,
    show_column_names   <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_row_names      <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_heatmap_legend <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-4-1.png" width="624" /></p>
</div>
<p>We can see nothing. The reason being is that most of the expression values were masked by some expression values that are extremely high. To explain this in a more intuitive way:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>gene_dist</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>normalised_counts</span>, MARGIN <span class='op'>=</span> <span class='fl'>1</span>, FUN <span class='op'>=</span> <span class='va'>hist</span>, plot <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>gene_col</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>rainbow</a></span><span class='op'>(</span>n <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>gene_dist</span><span class='op'>)</span>, alpha <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span>

<span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fl'>0</span>; <span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fl'>0</span>;
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>gene_dist</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>xlim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>xlim</span>, <span class='va'>gene_dist</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>$</span><span class='va'>breaks</span><span class='op'>)</span><span class='op'>)</span>
    <span class='va'>ylim</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/range.html'>range</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>ylim</span>, <span class='va'>gene_dist</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span><span class='op'>$</span><span class='va'>counts</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='co'>## Find the row that has the max count</span>
<span class='va'>gene_max_count</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>normalised_counts</span> <span class='op'>==</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>,
                        arr.ind <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>x        <span class='op'>=</span> <span class='va'>xlim</span>,
     y        <span class='op'>=</span> <span class='va'>ylim</span>,
     type     <span class='op'>=</span> <span class='st'>"n"</span>,
     main     <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Distributions for Expression Values of"</span>,
                      <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>, <span class='st'>"Genes"</span>,
                      <span class='st'>"in"</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>, <span class='st'>"Samples"</span><span class='op'>)</span>,
     cex.main <span class='op'>=</span> <span class='fl'>0.95</span>,
     xlab     <span class='op'>=</span> <span class='st'>"TMM normalised CPM"</span>,
     ylab     <span class='op'>=</span> <span class='st'>"Frequency"</span>
<span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_along</a></span><span class='op'>(</span><span class='va'>gene_dist</span><span class='op'>)</span><span class='op'>)</span><span class='op'>{</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>gene_dist</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>,
         col <span class='op'>=</span> <span class='va'>gene_col</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>,
         lty <span class='op'>=</span> <span class='st'>"blank"</span>,
         add <span class='op'>=</span> <span class='cn'>TRUE</span>
    <span class='op'>)</span>
<span class='op'>}</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topright"</span>,
       legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span><span class='op'>[</span><span class='va'>gene_max_count</span><span class='op'>]</span>,
       fill   <span class='op'>=</span> <span class='va'>gene_col</span><span class='op'>[</span><span class='va'>gene_max_count</span><span class='op'>]</span>,
       border <span class='op'>=</span> <span class='va'>gene_col</span><span class='op'>[</span><span class='va'>gene_max_count</span><span class='op'>]</span>,
       cex    <span class='op'>=</span> <span class='fl'>0.75</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-5-1.png" width="624" /></p>
</div>
<p>As shown by the distributions, while only a few genes have high expression values, the majority of the genes have counts far below 20000. A few extremely high expression values can mask most of the other expression values in the heatmap. This is why the heatmap above showing virtually all the genes have the same level of expression in the 86 samples. To show the difference in expression levels between genes, we need to transform the expression values for each gene to comparable scales.</p>
<h3 id="standardising-expression-values">Standardising expression values</h3>
<p>A common practice before ploting RNA-seq data in a heatmap is to standardise the normalised counts of a gene across all samples to z-scores, that is, calculate z-scores by genes(rows):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>z_mat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span>                    <span class='co'>## transpose matrix =&gt; row = samples, col = genes</span>
    <span class='fu'><a href='https://rdrr.io/r/base/scale.html'>scale</a></span><span class='op'>(</span>                    <span class='co'>## convert values of each column(gene) to Z scores</span>
        <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>
        <span class='op'>)</span>      <span class='co'>## transpose the transposed matrix =&gt; row =  genes, col = samples</span>
    <span class='op'>)</span>          
</code></pre>
</div>
</div>
<p>The heatmap after z-scoring:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>ComplexHeatmap</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html'>Heatmap</a></span><span class='op'>(</span>
    matrix              <span class='op'>=</span> <span class='va'>z_mat</span>,
    column_title        <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Expression Levels of"</span>,
                                <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span><span class='op'>)</span>,
                                <span class='st'>"Genes in 86 Samples of GSE152641"</span><span class='op'>)</span>,
    name                <span class='op'>=</span> <span class='st'>"Expr lvl"</span>,
    cluster_rows        <span class='op'>=</span> <span class='cn'>TRUE</span>,
    cluster_columns     <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_row_dend       <span class='op'>=</span> <span class='cn'>TRUE</span>,
    show_column_dend    <span class='op'>=</span> <span class='cn'>TRUE</span>,
    col                 <span class='op'>=</span> <span class='fu'>heatmap_cols</span><span class='op'>(</span><span class='va'>z_mat</span><span class='op'>)</span>,
    show_column_names   <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='co'>## Too many samples. Will be messy if show them</span>
    show_row_names      <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_heatmap_legend <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-7-1.png" width="624" /></p>
</div>
<p>The expression levels of the 14418 genes now become visible. However, we cannot tell which part of the heatmap corresponds to which sample, or more importantly which group of the 86 samples. Without this piece of information, it will be impossible to show the difference in gene expression level between the healthy control and the COVID-19 patients in the heatmap.</p>
<h3 id="heatmap-annotation">Heatmap annotation</h3>
<p>The <code>HeatmapAnnotation</code> object in <code>ComplexHeatmap</code> is dedicated to this purpose. The constructor method for <code>HeatmapAnnotation</code> requires two input arguments: a data frame with a column of group names for each sample and a list with a named vector of colour codes:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='co'>## Create a vector named with sample groups</span>
<span class='va'>groups</span>         <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span><span class='op'>)</span>    <span class='co'>## get group names of samples</span>
<span class='va'>ha_cols</span>        <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#B00020"</span>, <span class='st'>"#008B00"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>ha_cols</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>groups</span>
</code></pre>
</div>
</div>
<p>To present the annotation for the sample groups in a more organised way, we can re-arrange the column order of the count matrix by groups:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>group_order</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span> <span class='op'>==</span> <span class='va'>groups</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span>,
                 <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span> <span class='op'>==</span> <span class='va'>groups</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>    <span class='co'>## column index by groups</span>
</code></pre>
</div>
</div>
<p>And to create a heatmap annotation object using the constructor method:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>ha</span> <span class='op'>&lt;-</span> <span class='fu'>ComplexHeatmap</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html'>HeatmapAnnotation</a></span><span class='op'>(</span>
        df  <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>Sample <span class='op'>=</span> <span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>[</span><span class='va'>group_order</span><span class='op'>]</span><span class='op'>)</span>,
        col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Sample <span class='op'>=</span> <span class='va'>ha_cols</span><span class='op'>)</span>
    <span class='op'>)</span>
</code></pre>
</div>
</div>
<p>The heatmap with an annotation for the sample groups:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'>ComplexHeatmap</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html'>Heatmap</a></span><span class='op'>(</span>
    matrix              <span class='op'>=</span> <span class='va'>z_mat</span><span class='op'>[</span> , <span class='va'>group_order</span><span class='op'>]</span>,
    column_title        <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"Expression Levels of"</span>,
                                <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span><span class='op'>)</span>,
                                <span class='st'>"Genes in 86 Samples of GSE152641"</span><span class='op'>)</span>,
    name                <span class='op'>=</span> <span class='st'>"Expr lvl"</span>,
    cluster_rows        <span class='op'>=</span> <span class='cn'>TRUE</span>,
    cluster_columns     <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_row_dend       <span class='op'>=</span> <span class='cn'>TRUE</span>,
    show_column_dend    <span class='op'>=</span> <span class='cn'>TRUE</span>,
    col                 <span class='op'>=</span> <span class='fu'>heatmap_cols</span><span class='op'>(</span><span class='va'>z_mat</span><span class='op'>)</span>,
    show_column_names   <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='co'>## Too many samples. Will be messy if show them</span>
    show_row_names      <span class='op'>=</span> <span class='cn'>FALSE</span>,
    show_heatmap_legend <span class='op'>=</span> <span class='cn'>TRUE</span>,
    top_annotation      <span class='op'>=</span> <span class='va'>ha</span>      <span class='co'>## put annotation for sample groups at the top</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-11-1.png" width="624" /></p>
</div>
<p>Finally, putting everything above together, we can write a function to avoid duplicate code:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>plotHeatMap</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>deg</span>   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/integer.html'>integer</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>)</span>,
                        <span class='va'>m</span>,
                        <span class='va'>samples</span>,
                        <span class='va'>title</span> <span class='op'>=</span> <span class='st'>""</span>,
                        <span class='va'>cols</span>  <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#B00020"</span>, <span class='st'>"#008B00"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'># Purpose:</span>
    <span class='co'>#     Ad-hoc function to standardise values in the input matrix to Z-scores,</span>
    <span class='co'>#     and generate a heatmap plot for gene expression with samples ordered</span>
    <span class='co'>#     by their groups and annotated.</span>
    <span class='co'># Parameters:</span>
    <span class='co'>#     deg    : a vector of differentially expressed genes </span>
    <span class='co'>#     m      : RNA-seq count matrix</span>
    <span class='co'>#     samples:</span>
    <span class='co'>#              A dataframe specifying to which group a sample belongs.</span>
    <span class='co'>#              Note that it only supports samples divided into 2 groups:</span>
    <span class='co'>#              control and experimental</span>
    <span class='co'>#     title  : Title for the heatmap</span>
    <span class='co'>#     cols   :</span>
    <span class='co'>#              Colours to annotate the heatmap by groups.</span>
    <span class='co'>#              Use preset colours if not supplied.</span>
    <span class='co'># Value:</span>
    <span class='co'>#     gene_heatmap: A heatmap object</span>
    
    <span class='kw'><a href='https://rdrr.io/r/base/library.html'>require</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/jokergoo/ComplexHeatmap'>ComplexHeatmap</a></span><span class='op'>)</span>
    
    <span class='co'># Row normalisation/standardisation to Z-scores</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>deg</span><span class='op'>)</span> <span class='op'>==</span> <span class='fl'>0</span><span class='op'>)</span> <span class='op'>{</span>
        <span class='va'>m</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span>
            <span class='fu'><a href='https://rdrr.io/r/base/scale.html'>scale</a></span><span class='op'>(</span>
                <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>m</span><span class='op'>)</span>            <span class='co'># transpose matrix =&gt; row = samples, col = genes</span>
            <span class='op'>)</span>                  <span class='co'># convert values of each column(gene) to Z scores</span>
        <span class='op'>)</span>       <span class='co'># transpose the transposed matrix =&gt; row =  genes, col = samples</span>
    <span class='op'>}</span> <span class='kw'>else</span> <span class='op'>{</span>
        <span class='va'>m</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/scale.html'>scale</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>m</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>m</span><span class='op'>)</span> <span class='op'>%in%</span> <span class='va'>deg</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>
    <span class='op'>}</span>
    
    <span class='co'># Get group names of samples</span>
    <span class='va'>groups</span>      <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/levels.html'>levels</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span><span class='op'>)</span>
    <span class='co'># Sample order by group</span>
    <span class='va'>group_order</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span> <span class='op'>==</span> <span class='va'>groups</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>,
                     <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span> <span class='op'>==</span> <span class='va'>groups</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>
    
    <span class='co'># Annotate heatmap by groups</span>
    <span class='va'>ha_cols</span>        <span class='op'>&lt;-</span> <span class='va'>cols</span>
    <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>ha_cols</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='va'>groups</span>
    <span class='va'>ha</span> <span class='op'>&lt;-</span> <span class='fu'>ComplexHeatmap</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html'>HeatmapAnnotation</a></span><span class='op'>(</span>
        df  <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>Sample <span class='op'>=</span> <span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>[</span><span class='va'>group_order</span><span class='op'>]</span><span class='op'>)</span>,
        col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Sample <span class='op'>=</span> <span class='va'>ha_cols</span><span class='op'>)</span>
    <span class='op'>)</span>
    <span class='co'># Create a heatmap object</span>
    <span class='va'>gene_heatmap</span> <span class='op'>&lt;-</span> <span class='fu'>ComplexHeatmap</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html'>Heatmap</a></span><span class='op'>(</span>
        matrix              <span class='op'>=</span> <span class='va'>m</span><span class='op'>[</span> , <span class='va'>group_order</span><span class='op'>]</span>,
        column_title        <span class='op'>=</span> <span class='va'>title</span>,
        name                <span class='op'>=</span> <span class='st'>"Expr lvl"</span>,
        cluster_rows        <span class='op'>=</span> <span class='cn'>TRUE</span>,
        cluster_columns     <span class='op'>=</span> <span class='cn'>FALSE</span>,
        show_row_dend       <span class='op'>=</span> <span class='cn'>TRUE</span>,
        show_column_dend    <span class='op'>=</span> <span class='cn'>TRUE</span>,
        col                 <span class='op'>=</span> <span class='fu'>heatmap_cols</span><span class='op'>(</span><span class='va'>m</span><span class='op'>)</span>,
        show_column_names   <span class='op'>=</span> <span class='cn'>FALSE</span>,
        show_row_names      <span class='op'>=</span> <span class='cn'>FALSE</span>,
        show_heatmap_legend <span class='op'>=</span> <span class='cn'>TRUE</span>,
        top_annotation      <span class='op'>=</span> <span class='va'>ha</span>
    <span class='op'>)</span>
    
    <span class='kw'><a href='https://rdrr.io/r/base/function.html'>return</a></span><span class='op'>(</span><span class='va'>gene_heatmap</span><span class='op'>)</span>
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<h2 id="de-analysis-by-quasi-likelihood-methods-qlm">DE Analysis by Quasi-Likelihood Methods (QLM)</h2>
<h3 id="model-design">Model Design</h3>
<p>We hypothesise that COVID-19 status of samples was the only factor contributing to differential gene expression. Therefore, we model on <em>group</em> (status COVID-19 and healthy control) such that fitting this design matrix will tell us how the COVID-19 status of a sample explains his/her expression levels of the genes.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>model_design</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/model.matrix.html'>model.matrix</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>group</span>, data <span class='op'>=</span> <span class='va'>samples</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:200px; overflow-x: scroll; width:100%; ">
<table class=" lightable-paper" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
(Intercept)
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
groupHealthy control
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
GSM4622627
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622628
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622629
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622630
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622631
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622632
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622633
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622634
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622635
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622636
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622637
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622638
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622639
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622640
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622641
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622642
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622643
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622644
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622645
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622646
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622647
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622648
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622649
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622650
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622651
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622652
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622653
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622654
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622655
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622656
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622657
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622658
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622659
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622660
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622661
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622662
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622663
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622664
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622665
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622666
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622667
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622668
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622669
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622670
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622671
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622672
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622673
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622674
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622675
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622676
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622677
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622678
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622679
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622680
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622681
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622682
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622683
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622684
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622685
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622686
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622687
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622688
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622689
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622690
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622691
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622692
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622693
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622694
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622695
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622696
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622697
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622698
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622699
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622700
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622701
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622702
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622703
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622704
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622705
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622706
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622707
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622708
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622709
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622710
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622711
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622712
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that <em>samples</em> is a data frame labeling samples with their groups:</p>
<div class="layout-chunk" data-layout="l-body">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:200px; overflow-x: scroll; width:100%; ">
<table class=" lightable-paper" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
sample
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
group
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
GSM4622627
</td>
<td style="text-align:left;">
IMX_sample00001
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622628
</td>
<td style="text-align:left;">
IMX_sample00002
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622629
</td>
<td style="text-align:left;">
IMX_sample00003
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622630
</td>
<td style="text-align:left;">
IMX_sample00004
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622631
</td>
<td style="text-align:left;">
IMX_sample00005
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622632
</td>
<td style="text-align:left;">
IMX_sample00006
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622633
</td>
<td style="text-align:left;">
IMX_sample00007
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622634
</td>
<td style="text-align:left;">
IMX_sample00008
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622635
</td>
<td style="text-align:left;">
IMX_sample00009
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622636
</td>
<td style="text-align:left;">
IMX_sample00010
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622637
</td>
<td style="text-align:left;">
IMX_sample00011
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622638
</td>
<td style="text-align:left;">
IMX_sample00012
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622639
</td>
<td style="text-align:left;">
IMX_sample00013
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622640
</td>
<td style="text-align:left;">
IMX_sample00014
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622641
</td>
<td style="text-align:left;">
IMX_sample00015
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622642
</td>
<td style="text-align:left;">
IMX_sample00016
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622643
</td>
<td style="text-align:left;">
IMX_sample00017
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622644
</td>
<td style="text-align:left;">
IMX_sample00018
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622645
</td>
<td style="text-align:left;">
IMX_sample00019
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622646
</td>
<td style="text-align:left;">
IMX_sample00020
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622647
</td>
<td style="text-align:left;">
IMX_sample00021
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622648
</td>
<td style="text-align:left;">
IMX_sample00022
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622649
</td>
<td style="text-align:left;">
IMX_sample00023
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622650
</td>
<td style="text-align:left;">
IMX_sample00024
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622651
</td>
<td style="text-align:left;">
IMX_sample00025
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622652
</td>
<td style="text-align:left;">
IMX_sample00026
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622653
</td>
<td style="text-align:left;">
IMX_sample00027
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622654
</td>
<td style="text-align:left;">
IMX_sample00028
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622655
</td>
<td style="text-align:left;">
IMX_sample00029
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622656
</td>
<td style="text-align:left;">
IMX_sample00030
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622657
</td>
<td style="text-align:left;">
IMX_sample00031
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622658
</td>
<td style="text-align:left;">
IMX_sample00032
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622659
</td>
<td style="text-align:left;">
IMX_sample00033
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622660
</td>
<td style="text-align:left;">
IMX_sample00034
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622661
</td>
<td style="text-align:left;">
IMX_sample00035
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622662
</td>
<td style="text-align:left;">
IMX_sample00036
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622663
</td>
<td style="text-align:left;">
IMX_sample00037
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622664
</td>
<td style="text-align:left;">
IMX_sample00038
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622665
</td>
<td style="text-align:left;">
IMX_sample00039
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622666
</td>
<td style="text-align:left;">
IMX_sample00040
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622667
</td>
<td style="text-align:left;">
IMX_sample00041
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622668
</td>
<td style="text-align:left;">
IMX_sample00042
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622669
</td>
<td style="text-align:left;">
IMX_sample00043
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622670
</td>
<td style="text-align:left;">
IMX_sample00044
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622671
</td>
<td style="text-align:left;">
IMX_sample00045
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622672
</td>
<td style="text-align:left;">
IMX_sample00046
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622673
</td>
<td style="text-align:left;">
IMX_sample00047
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622674
</td>
<td style="text-align:left;">
IMX_sample00048
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622675
</td>
<td style="text-align:left;">
IMX_sample00049
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622676
</td>
<td style="text-align:left;">
IMX_sample00050
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622677
</td>
<td style="text-align:left;">
IMX_sample00051
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622678
</td>
<td style="text-align:left;">
IMX_sample00052
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622679
</td>
<td style="text-align:left;">
IMX_sample00053
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622680
</td>
<td style="text-align:left;">
IMX_sample00054
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622681
</td>
<td style="text-align:left;">
IMX_sample00055
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622682
</td>
<td style="text-align:left;">
IMX_sample00056
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622683
</td>
<td style="text-align:left;">
IMX_sample00057
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622684
</td>
<td style="text-align:left;">
IMX_sample00058
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622685
</td>
<td style="text-align:left;">
IMX_sample00059
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622686
</td>
<td style="text-align:left;">
IMX_sample00060
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622687
</td>
<td style="text-align:left;">
IMX_sample00061
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622688
</td>
<td style="text-align:left;">
IMX_sample00062
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622689
</td>
<td style="text-align:left;">
IMX_sample00063
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622690
</td>
<td style="text-align:left;">
IMX_sample00064
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622691
</td>
<td style="text-align:left;">
IMX_sample00065
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622692
</td>
<td style="text-align:left;">
IMX_sample00066
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622693
</td>
<td style="text-align:left;">
IMX_sample00067
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622694
</td>
<td style="text-align:left;">
IMX_sample00068
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622695
</td>
<td style="text-align:left;">
IMX_sample00069
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622696
</td>
<td style="text-align:left;">
IMX_sample00070
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622697
</td>
<td style="text-align:left;">
IMX_sample00071
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622698
</td>
<td style="text-align:left;">
IMX_sample00072
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622699
</td>
<td style="text-align:left;">
IMX_sample00073
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622700
</td>
<td style="text-align:left;">
IMX_sample00074
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622701
</td>
<td style="text-align:left;">
IMX_sample00075
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622702
</td>
<td style="text-align:left;">
IMX_sample00076
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622703
</td>
<td style="text-align:left;">
IMX_sample00077
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622704
</td>
<td style="text-align:left;">
IMX_sample00078
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622705
</td>
<td style="text-align:left;">
IMX_sample00079
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622706
</td>
<td style="text-align:left;">
IMX_sample00080
</td>
<td style="text-align:left;">
COVID19
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622707
</td>
<td style="text-align:left;">
IMX_sample00081
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622708
</td>
<td style="text-align:left;">
IMX_sample00082
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622709
</td>
<td style="text-align:left;">
IMX_sample00083
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622710
</td>
<td style="text-align:left;">
IMX_sample00084
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622711
</td>
<td style="text-align:left;">
IMX_sample00085
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
<tr>
<td style="text-align:left;">
GSM4622712
</td>
<td style="text-align:left;">
IMX_sample00086
</td>
<td style="text-align:left;">
Healthy control
</td>
</tr>
</tbody>
</table>
</div>
</div>
<h3 id="multiple-hypothesis-testing-for-differentially-expressed-genes-by-qlm">Multiple Hypothesis Testing for Differentially Expressed Genes by QLM</h3>
<p>We first use edgeR<span class="citation" data-cites="edgeR1">(<a href="#ref-edgeR1" role="doc-biblioref">Robinson, McCarthy, and Smyth 2009</a>)</span><span class="citation" data-cites="edgeR2">(<a href="#ref-edgeR2" role="doc-biblioref">McCarthy, Chen, and Smyth 2012</a>)</span> to estimate the dispersion. Note that edgeR stores data in a DGEList object and perform analyses on it:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>d</span>       <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>DGEList</a></span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>normalised_counts</span>,
                          group  <span class='op'>=</span> <span class='va'>samples</span><span class='op'>$</span><span class='va'>group</span><span class='op'>)</span>
<span class='va'>d</span>       <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/estimateDisp.html'>estimateDisp</a></span><span class='op'>(</span><span class='va'>d</span>, <span class='va'>model_design</span><span class='op'>)</span> <span class='co'># dispersion of normalised count</span>
<span class='va'>fit_qlm</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/glmQLFTest.html'>glmQLFit</a></span><span class='op'>(</span><span class='va'>d</span>, <span class='va'>model_design</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>In this step, QL dispersion was estimated with edgeR<span class="citation" data-cites="edgeR1">(<a href="#ref-edgeR1" role="doc-biblioref">Robinson, McCarthy, and Smyth 2009</a>)</span><span class="citation" data-cites="edgeR2">(<a href="#ref-edgeR2" role="doc-biblioref">McCarthy, Chen, and Smyth 2012</a>)</span> by fitting a QL negative binomial (NB) generalised linear model (glm) to the DGEList data containing the normalised counts and the estimated NB dispersions (visualised with <a href="https://www.ffli.dev/posts/2021-07-05-geo-datasets-experiment-selection-and-initial-processing/#biological-coefficient-of-variation-bcv">the BCV</a> and <a href="https://www.ffli.dev/posts/2021-07-05-geo-datasets-experiment-selection-and-initial-processing/#mean-variance-relationship">the mean-variance plot</a>), along with the design matrix we just created.</p>
<p>We then use the fitted result to conduct genewise QL F-test for the coefficients of our defined sample groups.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qlf_SARS2vsHC</span> <span class='op'>&lt;-</span> <span class='fu'>edgeR</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/glmQLFTest.html'>glmQLFTest</a></span><span class='op'>(</span><span class='va'>fit_qlm</span>, coef <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>model_design</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>This step calculates the p-values for each of the genes in our expression set.</p>
<h3 id="ql-f-test-p-values-comparison-fdr-correcred-vs-uncorrected">QL F-test P-values Comparison: FDR-correcred vs Uncorrected</h3>
<p>Extract the top differentially expressed tags ranked by p-values from the result of the QL F-test and display the number of genes that passed the correction:</p>
<div class="layout-chunk" data-layout="l-body">
<table>
<thead>
<tr>
<th style="text-align:right;">
Number of genes with p-value &lt; 0.05
</th>
<th style="text-align:right;">
Number of genes that pass after correction
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
7805
</td>
<td style="text-align:right;">
6856
</td>
</tr>
</tbody>
</table>
</div>
<p>The threshold for showing statistically significant evidence for differential expression was set to <span class="math inline">\(p &lt; 0.05\)</span> for each gene. This means genes that are captured only have less 5% chance to show such differences in expression between groups if they were non-differentially expressed genes.</p>
<p>Note that edgeR by default uses the false discovery rate (FDR/Benjamini-Hochberg) method to correct p-values for false positive results. This method was applied because we need to control for the the liklihood of false positive results that would increase by chance with the increasing number of tests performed. The threshold for the corrected p-value (FDR) is also set to be <span class="math inline">\(&lt;0.05\)</span> to capture genes that have false positive results for fewer than 5% of the significant tests.</p>
<h3 id="volcano-plot">Volcano Plot</h3>
<p>After hypothesis testing for detecting differentially expressed genes, it is of interest to compare the fold changes against the statistical significance levels. Commonly used to visualise this comparison for two treatment conditions is the volcano plot, where the horizontal dimension is the fold change between the two groups on the base-2 log scale, and the vertical dimension is the adjusted P-values on the negative base-10 log scale.</p>
<p>Another plot often used in interpreting the results of differential gene expression analysis is the MA plot, where the horizontal dimension is the mean of normalised gene counts (A) and the vertical dimension the fold changes, both on the base-2 log scale.<span class="citation" data-cites="DGEplot">(<a href="#ref-DGEplot" role="doc-biblioref">McDermaid et al. 2018</a>)</span> However, a volcano plot is a more preferable option to an MA plot for visualising DGE results, because an MA plot does not consider the association between the results of multiple hypothesis testing and the changes in expression levels between the two groups.</p>
<p>To create volcano plots with base R, we can write a simple function:</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>plotVolcano</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>df</span>, <span class='va'>title</span>, <span class='va'>gene_of_interest</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/integer.html'>integer</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'># Purpose:</span>
    <span class='co'>#     Plot a vocanol plot conveniently with highlighted genes of interest.</span>
    <span class='co'># Parameters:</span>
    <span class='co'>#     df: A dataframe dedicated to this dataset,</span>
    <span class='co'>#         must include the following ordered columns, regardless of names:</span>
    <span class='co'>#         column 1: HGNC gene symbol</span>
    <span class='co'>#         column 2: Fold change on log2 scale</span>
    <span class='co'>#         column 3: negated P-value on log10 scale</span>
    <span class='co'>#     title: title of the plot</span>
    <span class='co'>#     gene_of_interst:</span>
    <span class='co'>#         A vector of genes to highlight, optional</span>
    <span class='co'># Value:</span>
    <span class='co'>#     result: A volcano plot with genes of interest highlighted.</span>
    
    <span class='va'>cols</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"insig"</span>     <span class='op'>=</span> <span class='st'>"grey"</span>,
              <span class='st'>"sig"</span>       <span class='op'>=</span> <span class='st'>"#2cbe88"</span>,
              <span class='st'>"highlight"</span> <span class='op'>=</span> <span class='st'>"#d03b41"</span><span class='op'>)</span>
    
    <span class='va'>df</span><span class='op'>$</span><span class='va'>colour</span>         <span class='op'>&lt;-</span> <span class='va'>cols</span><span class='op'>[</span><span class='st'>"insig"</span><span class='op'>]</span>
    <span class='va'>sig</span>               <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>3</span><span class='op'>]</span> <span class='op'>&gt;</span> <span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='fl'>0.05</span><span class='op'>)</span><span class='op'>)</span>
    <span class='va'>df</span><span class='op'>[</span><span class='va'>sig</span>, <span class='st'>"colour"</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='va'>cols</span><span class='op'>[</span><span class='st'>"sig"</span><span class='op'>]</span>
    
    <span class='va'>highlight</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/integer.html'>integer</a></span><span class='op'>(</span><span class='fl'>0</span><span class='op'>)</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>gene_of_interest</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>)</span>
        <span class='va'>highlight</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>1</span><span class='op'>]</span> <span class='op'>%in%</span> <span class='va'>gene_of_interest</span><span class='op'>)</span>

    <span class='va'>x</span> <span class='op'>&lt;-</span> <span class='va'>df</span><span class='op'>[</span> ,<span class='fl'>2</span><span class='op'>]</span>
    <span class='va'>y</span> <span class='op'>&lt;-</span> <span class='va'>df</span><span class='op'>[</span> ,<span class='fl'>3</span><span class='op'>]</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>highlight</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>)</span>
        <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='st'>"colour"</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='va'>cols</span><span class='op'>[</span><span class='st'>"highlight"</span><span class='op'>]</span>
    
    <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>x    <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='op'>-</span><span class='va'>highlight</span>, <span class='fl'>2</span><span class='op'>]</span>,
         y    <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='op'>-</span><span class='va'>highlight</span>, <span class='fl'>3</span><span class='op'>]</span>,
         col  <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='op'>-</span><span class='va'>highlight</span>, <span class='st'>"colour"</span><span class='op'>]</span>,
         xlab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='st'>"Log"</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>*</span><span class='st'>" fold change"</span><span class='op'>)</span>,
         ylab <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='st'>"-Log"</span><span class='op'>[</span><span class='fl'>10</span><span class='op'>]</span><span class='op'>*</span><span class='st'>" P"</span><span class='op'>)</span>,
         main <span class='op'>=</span> <span class='va'>title</span><span class='op'>)</span>
    <span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>highlight</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>)</span> <span class='op'>{</span>
        <span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span>x      <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='fl'>2</span><span class='op'>]</span>,
               y      <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='fl'>3</span><span class='op'>]</span>,
               col    <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='st'>"colour"</span><span class='op'>]</span>,
               pch    <span class='op'>=</span> <span class='fl'>8</span>,
               cex    <span class='op'>=</span> <span class='fl'>1.5</span>,
               lwd    <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
        <span class='fu'><a href='https://rdrr.io/r/graphics/text.html'>text</a></span><span class='op'>(</span>x        <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='fl'>2</span><span class='op'>]</span>,
             y        <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='fl'>3</span><span class='op'>]</span>,
             labels   <span class='op'>=</span> <span class='va'>df</span><span class='op'>[</span><span class='va'>highlight</span>, <span class='fl'>1</span><span class='op'>]</span>,
             cex      <span class='op'>=</span> <span class='fl'>0.75</span><span class='op'>)</span>
        <span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>x      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>,
               y      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>3</span><span class='op'>]</span><span class='op'>)</span>,
               legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"has evidence of DE"</span>,
                          <span class='st'>"insignificant"</span>,
                          <span class='st'>"gene of interest"</span><span class='op'>)</span>,
               col    <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>cols</span><span class='op'>[</span><span class='st'>"sig"</span><span class='op'>]</span>, <span class='va'>cols</span><span class='op'>[</span><span class='st'>"insig"</span><span class='op'>]</span>, <span class='va'>cols</span><span class='op'>[</span><span class='st'>"highlight"</span><span class='op'>]</span><span class='op'>)</span>,
               pch    <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span>, <span class='fl'>8</span><span class='op'>)</span>,
               cex    <span class='op'>=</span> <span class='fl'>0.75</span><span class='op'>)</span>
        
    <span class='op'>}</span> <span class='kw'>else</span> <span class='op'>{</span>
        <span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>x      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>,
               y      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>3</span><span class='op'>]</span><span class='op'>)</span>,
               legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"has evidence of DE"</span>, <span class='st'>"insignificant"</span><span class='op'>)</span>,
               col    <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='va'>cols</span><span class='op'>[</span><span class='st'>"sig"</span><span class='op'>]</span>, <span class='va'>cols</span><span class='op'>[</span><span class='st'>"insig"</span><span class='op'>]</span><span class='op'>)</span>,
               cex    <span class='op'>=</span> <span class='fl'>0.75</span><span class='op'>)</span>
    <span class='op'>}</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/abline.html'>abline</a></span><span class='op'>(</span>h    <span class='op'>=</span> <span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='fl'>0.05</span><span class='op'>)</span>,
           col  <span class='op'>=</span> <span class='st'>"#0c82b7"</span>,
           lty  <span class='op'>=</span> <span class='fl'>2</span>,
           lwd  <span class='op'>=</span> <span class='fl'>1.5</span><span class='op'>)</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/text.html'>text</a></span><span class='op'>(</span>x      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span><span class='op'>(</span><span class='va'>df</span><span class='op'>[</span> , <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span> <span class='op'>+</span> <span class='fl'>0.25</span>,
         y      <span class='op'>=</span> <span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='fl'>0.05</span><span class='op'>)</span>  <span class='op'>+</span> <span class='fl'>0.5</span>,
         labels <span class='op'>=</span> <span class='st'>"p = 0.05"</span>,
         col    <span class='op'>=</span> <span class='st'>"#0c82b7"</span><span class='op'>)</span>
    
<span class='op'>}</span>
</code></pre>
</div>
</details>
</div>
<p>There are also packages that offer functions to create fancy volcano/MA plots such as <code>vidger</code>, but it is more fun to write one yourself:)</p>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>qlf_volcano</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    gene       <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>)</span>,
    log2FC     <span class='op'>=</span> <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>logFC</span>,
    neg_log10P <span class='op'>=</span> <span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>FDR</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='fu'>plotVolcano</span><span class='op'>(</span>
    df               <span class='op'>=</span> <span class='va'>qlf_volcano</span>,
    title            <span class='op'>=</span> <span class='st'>"QL F-test Results for 14418 Genes (HC vs COVID-19)"</span>,
    gene_of_interest <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ACO1"</span>, <span class='st'>"ATL3"</span>, <span class='st'>"ZC3H13"</span>, <span class='st'>"AMIGO1"</span><span class='op'>)</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-20-1.png" width="624" /></p>
</div>
<p>ACO1, ATL3, ZC3H13, and AMIGO1 are highlighted in the volcano plot. The reason is that the original study identified ACO1 and ATL3 to be completely discordant and oppositely regulated in COVID-19 and non-COVID-19 viral infections: ACO1 is over-expressed in COVID-19 and under-expressed in non-COVID19 infections, while ATL3 is under-expressed in COVID-19 and over-expressed in non-COVID-19 infections. The volcano plot is consistent with this observation made in the original study. As for ZC3H13 and AMIGO1, the author believes their expression may be inhibited by COVID-19 to escape the host response, and this is to be investigated in future studies.</p>
<h3 id="heatmap-for-qlf-top-hit-genes">Heatmap for QLF top-hit Genes</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>tophits_qlf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span>
  <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span>, <span class='op'>]</span>
  <span class='op'>)</span>

<span class='fu'>plotHeatMap</span><span class='op'>(</span>deg     <span class='op'>=</span> <span class='va'>tophits_qlf</span>,
            m       <span class='op'>=</span> <span class='va'>normalised_counts</span>,
            samples <span class='op'>=</span> <span class='va'>samples</span>,
            title   <span class='op'>=</span> <span class='st'>"Expression Levels of QLF Top-hit Genes"</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-21-1.png" width="624" /></p>
</div>
<p>The heatmap plotted from the top-hit genes (<span class="math inline">\(\mathrm{FDR}&lt; 0.05\)</span>) from the differential expression analysis exhibits clusters of up-regulated and down-regulated expressions between the two group. This was not observed in the heatmap plotted using all the genes because here we have removed genes that show no significant evidence for differential expression for COVID-19 status (<span class="math inline">\(\mathrm{FDR}\geq 0.05\)</span>), which introduced noises in the first heatmap.</p>
<h2 id="differential-expression-analysis-by-linear-models-of-mircroarray-limma">Differential Expression Analysis by Linear Models of MircroArray (Limma)</h2>
<p><a href="10.18129/B9.bioc.limma">Limma</a><span class="citation" data-cites="limma">(<a href="#ref-limma" role="doc-biblioref">Ritchie et al. 2015</a>)</span> has two methods for differential expression analysis: the precision weights approach (voom) and the empirical Bayes prior trend approach (limma-trend). The later limma-trend appraoch is simpler and more robust than voom, provided that the sequencing depth across the RNA samples is reasonably consistent. According to <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">the limma user guide</a>, the sequencing depth is considered to be reasonably consistent if the ratio of the largest library size to the smallest is not more than about 3-fold. The limma-trend method is applied here.<span class="citation" data-cites="limmaDEA">(<a href="#ref-limmaDEA" role="doc-biblioref">Phipson et al. 2016</a>)</span></p>
<h3 id="multiple-hypothesis-testing">Multiple Hypothesis Testing</h3>
<p>We are fitting the same model matrix as the one we used for the QLM fitting. Note that limma uses logCPM rather than CPM itself:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>fit_limma</span> <span class='op'>&lt;-</span> <span class='fu'>limma</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/limma/man/lmFit.html'>lmFit</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log2</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>,
                          design <span class='op'>=</span> <span class='va'>model_design</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Then we apply empircal Bayes to compute differential expression:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>fit_limma</span> <span class='op'>&lt;-</span> <span class='fu'>limma</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/limma/man/ebayes.html'>eBayes</a></span><span class='op'>(</span>
    fit   <span class='op'>=</span> <span class='va'>fit_limma</span>,
    trend <span class='op'>=</span> <span class='cn'>T</span> <span class='co'># specific to RNA-seq, as we are not working with microarray data</span>
<span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Get the top hits ranked by P-values:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>limma_output_hits</span> <span class='op'>&lt;-</span> <span class='fu'>limma</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/limma/man/toptable.html'>topTable</a></span><span class='op'>(</span>
    fit           <span class='op'>=</span> <span class='va'>fit_limma</span>,
    coef          <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>fit_limma</span><span class='op'>$</span><span class='va'>coefficients</span><span class='op'>)</span> <span class='op'>==</span> <span class='st'>"groupHealthy control"</span><span class='op'>)</span>,
    adjust.method <span class='op'>=</span> <span class='st'>"BH"</span>, <span class='co'># use Benjamni-Hochberg to correct for p-values</span>
    number        <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>normalised_counts</span><span class='op'>)</span>
    <span class='op'>)</span>
<span class='va'>limma_output_hits</span> <span class='op'>&lt;-</span> <span class='va'>limma_output_hits</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span><span class='op'>(</span><span class='va'>limma_output_hits</span><span class='op'>$</span><span class='va'>P.Value</span><span class='op'>)</span>, <span class='op'>]</span>
</code></pre>
</div>
</div>
<h3 id="p-value-adjustment-methods">P-value adjustment methods</h3>
<p>For the results of empircal Bayes, we are able to select whatever p-value correction mathods R offers:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>p.adjust.methods</span>
</code></pre>
</div>
<pre><code>[1] &quot;holm&quot;       &quot;hochberg&quot;   &quot;hommel&quot;     &quot;bonferroni&quot; &quot;BH&quot;        
[6] &quot;BY&quot;         &quot;fdr&quot;        &quot;none&quot;      </code></pre>
</div>
<p>Note that <em>BH</em> and <em>fdr</em> are synonymous; they both refer to the Benjamini-Hochberg procedure for FDR control. According to R documentation for these adjustment methods, <em>holm</em>, <em>hochberg</em>, <em>hommel</em>, and <em>bonferroni</em> are designed to control the family-wise error rate (the probability of making one or more false postive discoveries; Type 1 error) and they control the probability of making such error in the multi-hypotheses testing. <em>BH/fdr</em> and <em>BY</em> on the other hand allow false positive results, but control the proportion out of the significant tests. This means the latter two methods are less conservative and stringent than the first four. Although <em>hommel</em> is a more powerful method, the R documentation states that the difference between the results of these two methods are small while the p-values of <em>BH</em> are faster to calculate. Therefore the FDR method was also chosen here.</p>
<h3 id="limma-p-values-comparison-bh-adjusted-vs-unadjusted">Limma P-values Comparison: BH-adjusted vs Unadjusted</h3>
<p>Check the number of genes that have P-values below the 0.05 cutoff and genes that have Benjamni-Hochberg corrected P-values below the cutoff:</p>
<div class="layout-chunk" data-layout="l-body">
<table>
<thead>
<tr>
<th style="text-align:right;">
Number of genes with p-value &lt; 0.05
</th>
<th style="text-align:right;">
Number of genes that pass after correction
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
8222
</td>
<td style="text-align:right;">
7361
</td>
</tr>
</tbody>
</table>
</div>
<p>The result is similar to that of the QLM F-test, with more genes that passed the corrected p-value threshold.</p>
<h3 id="volcano-plot-for-limma-top-hit-genes">Volcano Plot for Limma top-hit Genes</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>limma_volcano</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    gene       <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>limma_output_hits</span><span class='op'>)</span>,
    log2FC     <span class='op'>=</span> <span class='va'>limma_output_hits</span><span class='op'>$</span><span class='va'>logFC</span>,
    neg_log10P <span class='op'>=</span> <span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log10</a></span><span class='op'>(</span><span class='va'>limma_output_hits</span><span class='op'>$</span><span class='va'>adj.P.Val</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='fu'>plotVolcano</span><span class='op'>(</span>
    df               <span class='op'>=</span> <span class='va'>limma_volcano</span>,
    title            <span class='op'>=</span> <span class='st'>"Limma-eBayes Results for 14418 Genes (HC)"</span>,
    gene_of_interest <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ACO1"</span>, <span class='st'>"ATL3"</span>, <span class='st'>"ZC3H13"</span>, <span class='st'>"AMIGO1"</span><span class='op'>)</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-27-1.png" width="624" /></p>
</div>
<p>The volcano plot is nearly identical to the one for the edgeR result.</p>
<h3 id="heatmap-for-limma-top-hit-genes">Heatmap for Limma top-hit Genes</h3>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>tophits_limma</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>limma_output_hits</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>limma_output_hits</span><span class='op'>$</span><span class='va'>adj.P.Val</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span>
    <span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span>
<span class='fu'>plotHeatMap</span><span class='op'>(</span>deg     <span class='op'>=</span> <span class='va'>tophits_limma</span>,
            m       <span class='op'>=</span> <span class='va'>normalised_counts</span>,
            title   <span class='op'>=</span> <span class='st'>"Expression Levels of Limma Top-hit Genes"</span>,
            samples <span class='op'>=</span> <span class='va'>samples</span><span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-28-1.png" width="624" /></p>
</div>
<p>The heatmap for the limma top-hit genes has the same pattern of clusters as the one for the QLF top-hit genes. This again is because we only retain genes that are shown to be significantly differentially expressed by limma-eBayes, and plotted the two groups side-by-side for comparison.</p>
<h2 id="result-comparing-ql-f-test-and-limma-results">Result: Comparing QL F-test and Limma Results</h2>
<div class="layout-chunk" data-layout="l-body">
<details>
<summary>
Show code
</summary>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mfrow <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># Scatter plots for un-adjusted p-values ======================================</span>
<span class='va'>qlf_pvalues</span>     <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    gene_name      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>)</span>,
    qlf_pvalue     <span class='op'>=</span> <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>PValue</span>,
    qlf_pvalue_adj <span class='op'>=</span> <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>FDR</span>
    <span class='op'>)</span>
<span class='va'>limma_pvalues</span>     <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    gene_name        <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>limma_output_hits</span><span class='op'>)</span>,
    limma_pvalue     <span class='op'>=</span> <span class='va'>limma_output_hits</span><span class='op'>$</span><span class='va'>P.Value</span>,
    limma_pvalue_adj <span class='op'>=</span> <span class='va'>limma_output_hits</span><span class='op'>$</span><span class='va'>adj.P.Val</span>
    <span class='op'>)</span>
<span class='va'>qlf_limma_pvalues</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/merge.html'>merge</a></span><span class='op'>(</span>
    x    <span class='op'>=</span> <span class='va'>qlf_pvalues</span>,
    y    <span class='op'>=</span> <span class='va'>limma_pvalues</span>,
    by.x <span class='op'>=</span> <span class='st'>"gene_name"</span>,
    by.y <span class='op'>=</span> <span class='st'>"gene_name"</span>,
<span class='op'>)</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span> <span class='op'>&lt;-</span> <span class='st'>"grey"</span>
<span class='co'># Colour for QLF p-values below cutoff</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span><span class='op'>[</span>
    <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>qlf_pvalue</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"#92d964"</span>
<span class='co'># colours for limma p-values below cutoff</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span><span class='op'>[</span>
    <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>limma_pvalue</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"#29a2ca"</span>
<span class='co'># Colour for intersection of QLF and limma p values below cutoff</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>qlf_pvalue</span>   <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span>
    <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>limma_pvalue</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"#fae213"</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
    x        <span class='op'>=</span> <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>limma_pvalue</span>,
    y        <span class='op'>=</span> <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>qlf_pvalue</span>,
    col      <span class='op'>=</span> <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span>,
    xlab     <span class='op'>=</span> <span class='st'>"Limma Unadjusted P-values"</span>,
    ylab     <span class='op'>=</span> <span class='st'>"QLF Unadjusted P-values"</span>,
    cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>,
    main     <span class='op'>=</span> <span class='st'>"QLF vs Limma for Unadjusted P-values"</span>,
    cex.main <span class='op'>=</span> <span class='fl'>0.75</span>
<span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>
    x      <span class='op'>=</span> <span class='fl'>0</span>,
    y      <span class='op'>=</span> <span class='fl'>1</span>,
    legend <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"QLF"</span>, <span class='st'>"Limma"</span>, <span class='st'>"Both"</span>, <span class='st'>"not significant"</span><span class='op'>)</span>,
    fill   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#92d964"</span>, <span class='st'>"#29a2ca"</span>, <span class='st'>"#fae213"</span>, <span class='st'>"grey"</span><span class='op'>)</span>,
    cex    <span class='op'>=</span> <span class='fl'>0.5</span>
<span class='op'>)</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span> <span class='op'>&lt;-</span> <span class='st'>"grey"</span>
<span class='co'># Scatter plots for adjusted p-values =========================================</span>
<span class='co'># Colour for FDR-corrected QLF p-values below cutoff</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>qlf_pvalue_adj</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"#168c6e"</span>
<span class='co'># Colour for BH-adjusted limma p-values below cutoff</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>limma_pvalue_adj</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"#234eaa"</span>
<span class='co'># Colour for intersection of corrected QLF and limma p values below cutoff</span>
<span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>limma_pvalue_adj</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span>
    <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>qlf_pvalue_adj</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span><span class='op'>)</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='st'>"#cb641d"</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span>
    x        <span class='op'>=</span> <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>limma_pvalue_adj</span>,
    y        <span class='op'>=</span> <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>qlf_pvalue_adj</span>,
    col      <span class='op'>=</span> <span class='va'>qlf_limma_pvalues</span><span class='op'>$</span><span class='va'>colour</span>,
    xlab     <span class='op'>=</span> <span class='st'>"Limma BH-corrected P-values"</span>,
    ylab     <span class='op'>=</span> <span class='st'>"QLF FDR-corrected P-values"</span>,
    cex.lab  <span class='op'>=</span> <span class='fl'>0.75</span>,
    main     <span class='op'>=</span> <span class='st'>"QLF vs Limma for Adjusted P-values"</span>,
    cex.main <span class='op'>=</span> <span class='fl'>0.75</span>
<span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span>
    x        <span class='op'>=</span> <span class='fl'>0</span>,
    y        <span class='op'>=</span> <span class='fl'>1</span>,
    legend   <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"QLF"</span>, <span class='st'>"Limma"</span>, <span class='st'>"Both"</span>, <span class='st'>"not significant"</span><span class='op'>)</span>,
    fill     <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#168c6e"</span>, <span class='st'>"#234eaa"</span>, <span class='st'>"#cb641d"</span>, <span class='st'>"grey"</span><span class='op'>)</span>,
    cex      <span class='op'>=</span> <span class='fl'>0.5</span>
<span class='op'>)</span>
</code></pre>
</div>
</details>
<p><img src="differential-expression-analysis-edger-vs-limma_files/figure-html5/unnamed-chunk-29-1.png" width="624" /></p>
</div>
<p>The p-values calculated by QLF and Limma methods appear to be patternless. These two methods only capture a small common subset of differentially expressed genes. A question one might ask is, which method is better to use. Unfortunately this is also a question that I have not been able to answer. For RNA-seq differential expression analysis, I prefer edgeR to limma as limma is mostly used for microarray analyses, and the RNA-seq data is analysed as if it was microarray data by limma.</p>
<h2 id="to-do">To Do</h2>
<p>In later posts, enrichment or over-representation analysis (ORA) will be performed based on the DE analysis result by edgeR. Before performing the enrichment analysis, we need to create thresholded lists of genes for thresholded ORA and a ranked gene list for gene set enrichment analysis (GSEA).</p>
<h3 id="thresholded-list">Thresholded List</h3>
<p>We need to create 2 thresholded lists: one for the differentially expressed genes that were up-regulated, and one for those downregulated, with a threshold of 0.05 for the FDR-corrected p-value.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>upregulated_genes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&gt;</span> <span class='fl'>0</span>
<span class='op'>)</span>,<span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>x         <span class='op'>=</span> <span class='va'>upregulated_genes</span>,
            file      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"covid19_upregulated_genes.txt"</span><span class='op'>)</span>,
            sep       <span class='op'>=</span> <span class='st'>"\t"</span>,
            row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
            col.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
            quote     <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>downregulated_genes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span>
    <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>FDR</span> <span class='op'>&lt;</span> <span class='fl'>0.05</span> <span class='op'>&amp;</span> <span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>logFC</span> <span class='op'>&lt;</span> <span class='fl'>0</span>
<span class='op'>)</span>,<span class='op'>]</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>x         <span class='op'>=</span> <span class='va'>downregulated_genes</span>,
            file      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"covid19_downregulated_genes.txt"</span><span class='op'>)</span>,
            sep       <span class='op'>=</span> <span class='st'>"\t"</span>,
            row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
            col.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
            quote     <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<h3 id="ranked-list">Ranked List</h3>
<p>A non-thresholded, ranked list of genes for GSEA:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span class='va'>gene_rank</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
    genename <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>)</span>,
    F_stat   <span class='op'>=</span> <span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>FDR</span>,
                    base <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/sign.html'>sign</a></span><span class='op'>(</span><span class='va'>qlf_output_hits</span><span class='op'>$</span><span class='va'>table</span><span class='op'>$</span><span class='va'>logFC</span><span class='op'>)</span>
<span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span>x         <span class='op'>=</span> <span class='va'>gene_rank</span>,
            file      <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"."</span>, <span class='st'>"data"</span>, <span class='st'>"covid19_ranked_genelist.txt"</span><span class='op'>)</span>,
            sep       <span class='op'>=</span> <span class='st'>"\t"</span>,
            row.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
            col.names <span class='op'>=</span> <span class='cn'>FALSE</span>,
            quote     <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</code></pre>
</div>
</div>
<p>Note that we define the rank to be <span class="math inline">\(-\log_{10}{\mathrm{FDR}} \ \cdot\ \operatorname{sgn}(\log{\mathrm{FC}})\)</span>. With the <span class="math inline">\(-\log_{10}{\mathrm{FDR}}\)</span>, the samller the (corrected) p-value for a gene is, the higher the rank of that gene, regardless of whether it is up- or downregulated. The other term <span class="math inline">\(\operatorname{sgn}(\log{\mathrm{FC}})\)</span> take into account the factor of regulation of gene expression: if it is up-regulated, then it will be ranked from the top; if it is down-regulated, then it will be ranked from the bottom. Thereby we have genes that show the most statistically significant evidence of differential expression at the top for those upregulated, and the bottom of the list for those downregulated, while genes in the middle are least significant.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-complexheatmap" class="csl-entry" role="doc-biblioentry">
Gu, Zuguang, Roland Eils, and Matthias Schlesner. 2016. <span><span class="nocase">Complex heatmaps reveal patterns and correlations in multidimensional genomic data</span>.</span> <em>Bioinformatics</em> 32 (18): 284749. <a href="https://doi.org/10.1093/bioinformatics/btw313">https://doi.org/10.1093/bioinformatics/btw313</a>.
</div>
<div id="ref-circlize" class="csl-entry" role="doc-biblioentry">
Gu, Zuguang, Lei Gu, Roland Eils, Matthias Schlesner, and Benedikt Brors. 2014. <span><span class="nocase">circlize implements and enhances circular visualization in R </span>.</span> <em>Bioinformatics</em> 30 (19): 281112. <a href="https://doi.org/10.1093/bioinformatics/btu393">https://doi.org/10.1093/bioinformatics/btu393</a>.
</div>
<div id="ref-edgeR2" class="csl-entry" role="doc-biblioentry">
McCarthy, Davis J., Yunshun Chen, and Gordon K. Smyth. 2012. <span><span class="nocase">Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation</span>.</span> <em>Nucleic Acids Research</em> 40 (10): 428897. <a href="https://doi.org/10.1093/nar/gks042">https://doi.org/10.1093/nar/gks042</a>.
</div>
<div id="ref-DGEplot" class="csl-entry" role="doc-biblioentry">
McDermaid, Adam, Brandon Monier, Jing Zhao, Bingqiang Liu, and Qin Ma. 2018. <span><span class="nocase">Interpretation of differential gene expression results of RNA-seq data: review and integration</span>.</span> <em>Briefings in Bioinformatics</em> 20 (6): 204454. <a href="https://doi.org/10.1093/bib/bby067">https://doi.org/10.1093/bib/bby067</a>.
</div>
<div id="ref-limmaDEA" class="csl-entry" role="doc-biblioentry">
Phipson, Belinda, Stanley Lee, Ian J. Majewski, Warren S. Alexander, and Gordon K. Smyth. 2016. <span><span class="nocase">Robust hyperparameter estimation protects against hypervariable genes and improves power to detect differential expression</span>.</span> <em>The Annals of Applied Statistics</em> 10 (2): 94663. <a href="https://doi.org/10.1214/16-AOAS920">https://doi.org/10.1214/16-AOAS920</a>.
</div>
<div id="ref-limma" class="csl-entry" role="doc-biblioentry">
Ritchie, Matthew E., Belinda Phipson, Di Wu, Yifang Hu, Charity W. Law, Wei Shi, and Gordon K. Smyth. 2015. <span><span class="nocase">limma powers differential expression analyses for RNA-sequencing and microarray studies</span>.</span> <em>Nucleic Acids Research</em> 43 (7): e4747. <a href="https://doi.org/10.1093/nar/gkv007">https://doi.org/10.1093/nar/gkv007</a>.
</div>
<div id="ref-edgeR1" class="csl-entry" role="doc-biblioentry">
Robinson, Mark D., Davis J. McCarthy, and Gordon K. Smyth. 2009. <span><span class="nocase">edgeR: a Bioconductor package for differential expression analysis of digital gene expression data</span>.</span> <em>Bioinformatics</em> 26 (1): 13940. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-GSE152641" class="csl-entry" role="doc-biblioentry">
Thair, Simone A., Yudong D. He, Yehudit Hasin-Brumshtein, Suraj Sakaram, Rushika Pandya, Jiaying Toh, David Rawling, et al. 2021. <span>Transcriptomic Similarities and Differences in Host Response Between SARS-CoV-2 and Other Viral Infections.</span> <em>iScience</em> 24 (1): 101947. https://doi.org/<a href="https://doi.org/10.1016/j.isci.2020.101947">https://doi.org/10.1016/j.isci.2020.101947</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
